{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "db496b92",
   "metadata": {},
   "source": [
    "# Run Ablation Test on *SDePER* on simulated data: Scenario 1 + scRNA-seq data as reference + NO CVAE + WITHOUT sparsity penalty + WITHOUT spatial correlation constraint\n",
    "\n",
    "In this Notebook we run **ablation test** on SDePER on simulated data. For generating simulated data via coarse-graining procedure please refer [generate_simulated_spatial_data.nb.html](https://rawcdn.githack.com/az7jh2/SDePER_Analysis/c60dcb036816bd61b5a8b3752d473a5b591b52b6/Simulation/Generate_simulation_data/generate_simulated_spatial_data.nb.html).\n",
    "\n",
    "**Scenario 1** means the reference data for deconvolution includes all single cells with the **matched 12 cell types**.\n",
    "\n",
    "**scRNA-seq data as reference** means the reference data is scRNA-seq data ([GSE115746](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE115746)) from the same tissue with simulated spatial data, therefore **platform effect exists**.\n",
    "\n",
    "**NO CVAE** means we DO NOT use CVAE to remove platform effect although platform effect exists here.\n",
    "\n",
    "**WITHOUT sparsity penalty** means we fit the graph Laplacian regularized model without the sparsity penalty, specifically omitting Adaptive LASSO regularization (by setting the command option `--lambda_r` to 0).\n",
    "\n",
    "**WITHOUT spatial correlation constraint** means we fit the graph Laplacian regularized model the spatial correlation constraint, essentially removing Laplacian regularization (by setting the command option `--lambda_g` to 0)\n",
    "\n",
    "==================================================================================================================\n",
    "\n",
    "So here we use the **4 input files** as shown below:\n",
    "\n",
    "1. raw nUMI counts of simulated spatial transcriptomic data (spots × genes): [sim_spatial_spot_nUMI.csv](https://github.com/az7jh2/SDePER_Analysis/blob/main/Simulation/Generate_simulation_data/sim_spatial_spot_nUMI.csv)\n",
    "2. raw nUMI counts of reference scRNA-seq data (cells × genes): `scRNA_data_full.csv`. Since the file size of csv file of raw nUMI matrix of all 23,178 cells and 45,768 genes is up to 2.3 GB, we do not provide this file in our repository. It's just a **matrix transpose** of [GSE115746_cells_exon_counts.csv.gz](https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE115746&format=file&file=GSE115746%5Fcells%5Fexon%5Fcounts%2Ecsv%2Egz) in [GSE115746](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE115746) to satisty the file format requirement that rows as cells and columns as genes.\n",
    "3. cell type annotations for cells of **the matched 12 cell types** in reference scRNA-seq data (cells × 1): [ref_scRNA_cell_celltype.csv](https://github.com/az7jh2/SDePER_Analysis/blob/main/Simulation/Run_SDePER_on_simulation_data/Scenario_1/ref_scRNA_seq/ref_scRNA_cell_celltype.csv)\n",
    "4. adjacency matrix of spots in simulated spatial transcriptomic data (spots × spots): [sim_spatial_spot_adjacency_matrix.csv](https://github.com/az7jh2/SDePER_Analysis/blob/main/Simulation/Generate_simulation_data/sim_spatial_spot_adjacency_matrix.csv)\n",
    "\n",
    "==================================================================================================================\n",
    "\n",
    "SDePER settings are the same as baseline run [S1_ref_scRNA_SDePER_WITH_CVAE.ipynb](https://github.com/az7jh2/SDePER_Analysis/blob/main/Simulation/Run_SDePER_on_simulation_data/Scenario_1/ref_scRNA_seq/S1_ref_scRNA_SDePER_WITH_CVAE.ipynb):\n",
    "\n",
    "* number of included highly variable genes `n_hv_gene`: 500\n",
    "* number of selected TOP marker genes for each comparison in Differential `n_marker_per_cmp`: 50\n",
    "* seed for random values `seed`: 2\n",
    "* number of used CPU cores `n_core`: 64\n",
    "\n",
    "ALL other options are left as default.\n",
    "\n",
    "**For ablation test, disable CVAE, set hyper-parameter for Adaptive Lasso `lambda_r` as 0, and hyper-parameter for Graph Laplacian Constrain `lambda_g` as 0**.\n",
    "\n",
    "==================================================================================================================\n",
    "\n",
    "the `bash` command to start cell type deconvolution is\n",
    "\n",
    "`runDeconvolution -q sim_spatial_spot_nUMI.csv -r scRNA_data_full.csv -c ref_scRNA_cell_celltype.csv -a sim_spatial_spot_adjacency_matrix.csv --n_hv_gene 500 --n_marker_per_cmp 50 --seed 2 -n 64 --use_cvae false --lambda_r 0 --lambda_g 0`\n",
    "\n",
    "Note this Notebook uses **SDePER v1.0.0**. Cell type deconvolution result is renamed as [S1_ref_scRNA_SDePER_Ablation_NO_CVAE_LASSO_Laplacian_celltype_proportions.csv](https://github.com/az7jh2/SDePER_Analysis/blob/main/Ablation/Ablation_simulation_data/S1_ref_scRNA_SDePER_Ablation_NO_CVAE_LASSO_Laplacian_celltype_proportions.csv)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "df9e3939",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "SDePER (Spatial Deconvolution method with Platform Effect Removal) v1.0.0\n",
      "\n",
      "\n",
      "running options:\n",
      "spatial_file: /home/exouser/Spatial/sim_spatial_spot_nUMI.csv\n",
      "ref_file: /home/exouser/Spatial/scRNA_data_full.csv\n",
      "ref_celltype_file: /home/exouser/Spatial/ref_scRNA_cell_celltype.csv\n",
      "marker_file: None\n",
      "loc_file: None\n",
      "A_file: /home/exouser/Spatial/sim_spatial_spot_adjacency_matrix.csv\n",
      "n_cores: 64\n",
      "lambda_r: 0.0\n",
      "lambda_g: 0.0\n",
      "use_cvae: False\n",
      "threshold: 0\n",
      "n_hv_gene: 500\n",
      "n_marker_per_cmp: 50\n",
      "pseudo_spot_min_cell: 2\n",
      "pseudo_spot_max_cell: 8\n",
      "seq_depth_scaler: 10000\n",
      "cvae_input_scaler: 10\n",
      "cvae_init_lr: 0.003\n",
      "redo_de: True\n",
      "seed: 2\n",
      "diagnosis: False\n",
      "verbose: True\n",
      "use_imputation: False\n",
      "diameter: 200\n",
      "impute_diameter: [160, 114, 80]\n",
      "\n",
      "\n",
      "######### Preprocessing... #########\n",
      "\n",
      "######### Building CVAE skipped... #########\n",
      "\n",
      "read spatial data from file /home/exouser/Spatial/sim_spatial_spot_nUMI.csv\n",
      "total 581 spots; 1020 genes\n",
      "\n",
      "no marker gene profile provided. Perform DE to get cell-type marker genes on scRNA-seq data...\n",
      "\n",
      "read scRNA-seq data from file /home/exouser/Spatial/scRNA_data_full.csv\n",
      "total 23178 cells; 45768 genes\n",
      "read scRNA-seq cell-type annotation from file /home/exouser/Spatial/ref_scRNA_cell_celltype.csv\n",
      "total 12 cell-types\n",
      "subset cells with cell-type annotation, finally keep 11835 cells; 45768 genes\n",
      "\n",
      "Differential analysis across cell-types on scRNA-seq data...\n",
      "finally selected 677 cell-type marker genes\n",
      "\n",
      "\n",
      "gene filtering before modeling...\n",
      "all genes passed filtering\n",
      "\n",
      "spot filtering before modeling...\n",
      "all spots passed filtering\n",
      "\n",
      "\n",
      "######### Start GLRM modeling... #########\n",
      "\n",
      "GLRM settings:\n",
      "use SciPy minimize method:  L-BFGS-B\n",
      "global optimization turned off, local minimum will be used in GLRM\n",
      "use hybrid version of GLRM\n",
      "Numba detected total 64 available CPU cores. Use 64 CPU cores\n",
      "use 2001 points to calculate the heavy-tail density\n",
      "use weight threshold for Adaptive Lasso:  0.001\n",
      "total 117 unique nUMIs, min: 0, max: 267\n",
      "\n",
      "Build graph: \n",
      " Graph with 581 nodes and 1029 edges\n",
      "\n",
      "Estimate gene-specific platform effect gamma_g...\n",
      "estimate platform effect gammg_g by L-BFGS-B and basinhopping\n",
      "\n",
      "GLRM model initialization...\n",
      "CAUTION: global optimization turned on!\n",
      "calculate MLE theta and sigma^2 iteratively...\n",
      "  iter | time_opt | time_sig | sigma2\n",
      "     0 |    1.604 |    0.130 |  1.711\n",
      "     1 |    0.852 |    0.140 |  1.714\n",
      "     2 |    0.668 |    0.120 |  1.753\n",
      "     3 |    0.845 |    0.123 |  1.756\n",
      "     4 |    0.649 |    0.119 |  1.754\n",
      "     5 |    0.841 |    0.174 |  1.754\n",
      "MLE theta and sigma^2 calculation finished. Elapsed time: 0.10 minutes.\n",
      "gamma_g estimation finished. Elapsed time: 0.10 minutes.\n",
      "\n",
      "\n",
      "Start GLRM fitting...\n",
      "\n",
      "first estimate MLE theta and corresponding e^alpha and sigma^2...\n",
      "\n",
      "GLRM model initialization...\n",
      "calculate MLE theta and sigma^2 iteratively...\n",
      "  iter | time_opt | time_sig | sigma2\n",
      "     0 |   36.061 |    3.837 |  0.423\n",
      "     1 |   28.280 |    2.961 |  0.262\n",
      "     2 |   23.492 |    6.214 |  0.217\n",
      "     3 |   18.612 |    3.029 |  0.205\n",
      "     4 |   12.928 |    2.877 |  0.202\n",
      "     5 |    9.068 |    2.402 |  0.201\n",
      "MLE theta and sigma^2 calculation finished. Elapsed time: 2.50 minutes.\n",
      "MLE theta estimation finished. Elapsed time: 2.50 minutes.\n",
      "\n",
      "calculate weights of Adaptive Lasso...\n",
      "\n",
      "Stage 1: variable selection using Adaptive Lasso starts with the MLE theta and e^alpha, using already estimated sigma^2 and gamma_g...\n",
      "specified hyper-parameter for Adaptive Lasso is: 0.0\n",
      "\n",
      "start ADMM iteration...\n",
      "  iter |  res_pri_n | res_dual_n |    eps_pri |   eps_dual |        rho |    new_rho | time_opt | time_reg | time_lap | tilde_RMSE |   hat_RMSE\n",
      "     0 |      0.002 |      0.002 |      0.118 |      0.118 |       1.00 |          / |    5.105 |    0.000 |    0.004 |   0.000026 |   0.000013\n",
      "Terminated (optimal) in 1 iterations.\n",
      "One optimization by ADMM finished. Elapsed time: 0.09 minutes.\n",
      "\n",
      "Stage 1 variable selection finished. Elapsed time: 0.09 minutes.\n",
      "\n",
      "Stage 2: final theta estimation with Graph Laplacian Constrain using already estimated sigma^2 and gamma_g\n",
      "specified hyper-parameter for Graph Laplacian Constrain is: 0.0\n",
      "\n",
      "start ADMM iteration...\n",
      "  iter |  res_pri_n | res_dual_n |    eps_pri |   eps_dual |        rho |    new_rho | time_opt | time_reg | time_lap | tilde_RMSE |   hat_RMSE\n",
      "     0 |      4.783 |      4.783 |      0.123 |      0.123 |       1.00 |       1.00 |   34.549 |    0.000 |    0.003 |   0.079902 |   0.039951\n",
      "     1 |      0.001 |      4.783 |      0.123 |      0.123 |       1.00 |          / |    2.345 |    0.000 |    0.003 |   0.000002 |   0.000001\n",
      "early stop!\n",
      "Terminated (optimal) in 2 iterations.\n",
      "One optimization by ADMM finished. Elapsed time: 0.62 minutes.\n",
      "\n",
      "\n",
      "stage 2 finished. Elapsed time: 0.62 minutes.\n",
      "\n",
      "GLRM fitting finished. Elapsed time: 3.20 minutes.\n",
      "\n",
      "\n",
      "Post-processing estimated cell-type proportion theta...\n",
      "hard thresholding small theta values with threshold 0\n",
      "\n",
      "\n",
      "cell type deconvolution finished. Estimate results saved in /home/exouser/Spatial/celltype_proportions.csv. Elapsed time: 0.10 hours.\n",
      "\n",
      "\n",
      "######### No imputation #########\n",
      "\n",
      "\n",
      "whole pipeline finished. Total elapsed time: 0.10 hours.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "CompletedProcess(args='runDeconvolution -q sim_spatial_spot_nUMI.csv                           -r scRNA_data_full.csv                           -c ref_scRNA_cell_celltype.csv                           -a sim_spatial_spot_adjacency_matrix.csv                           --n_hv_gene 500                           --n_marker_per_cmp 50                           --seed 2                           -n 64                           --use_cvae false                           --lambda_r 0                           --lambda_g 0\\n', returncode=0)"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import subprocess\n",
    "\n",
    "cmd = '''runDeconvolution -q sim_spatial_spot_nUMI.csv \\\n",
    "                          -r scRNA_data_full.csv \\\n",
    "                          -c ref_scRNA_cell_celltype.csv \\\n",
    "                          -a sim_spatial_spot_adjacency_matrix.csv \\\n",
    "                          --n_hv_gene 500 \\\n",
    "                          --n_marker_per_cmp 50 \\\n",
    "                          --seed 2 \\\n",
    "                          -n 64 \\\n",
    "                          --use_cvae false \\\n",
    "                          --lambda_r 0 \\\n",
    "                          --lambda_g 0\n",
    "'''\n",
    "\n",
    "subprocess.run(cmd, check=True, text=True, shell=True)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
