{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "02f03238",
   "metadata": {},
   "source": [
    "# Run *SDePER* on simulated data: Platform effect demonstration + scRNA-seq data as reference + NO platform effect removal\n",
    "\n",
    "In this Notebook we run SDePER on STARmap-based simulated data. To best demonstrate the platform effect, we treat all 2,002 STARmap single cells as fake spatial spots and performed cell type deconvolution.\n",
    "\n",
    "**Platform effect demonstration** means we actually use single cells for cell type deconvolution to best demonstrate the platform effect. The reference data for deconvolution includes all single cells with the **matched 12 cell types**.\n",
    "\n",
    "**scRNA-seq data as reference** means the reference data is scRNA-seq data ([GSE115746](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE115746)) from the same tissue with simulated spatial data, therefore **platform effect exists**.\n",
    "\n",
    "**NO platform effect removal** means we conduct cell type deconvolution **disregarding platform effect**, meaning that neither CVAE nor an additive gene-wise platform effect term is utilized.\n",
    "\n",
    "==================================================================================================================\n",
    "\n",
    "So here we use the **3 input files** as shown below:\n",
    "\n",
    "1. raw nUMI counts of all 2,002 STARmap single cells (cells × genes): [STARmap_cell_nUMI.csv](https://github.com/az7jh2/SDePER_Analysis/blob/main/Simulation/Run_SDePER_on_simulation_data/Scenario_1/ref_spatial/STARmap_cell_nUMI.csv)\n",
    "2. raw nUMI counts of reference scRNA-seq data (cells × genes): `scRNA_data_full.csv`. Since the file size of csv file of raw nUMI matrix of all 23,178 cells and 45,768 genes is up to 2.3 GB, we do not provide this file in our repository. It's just a **matrix transpose** of [GSE115746_cells_exon_counts.csv.gz](https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE115746&format=file&file=GSE115746%5Fcells%5Fexon%5Fcounts%2Ecsv%2Egz) in [GSE115746](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE115746) to satisty the file format requirement that rows as cells and columns as genes.\n",
    "3. cell type annotations for cells of **the matched 12 cell types** in reference scRNA-seq data (cells × 1): [ref_scRNA_cell_celltype.csv](https://github.com/az7jh2/SDePER_Analysis/blob/main/Simulation/Run_SDePER_on_simulation_data/Scenario_1/ref_scRNA_seq/ref_scRNA_cell_celltype.csv)\n",
    "\n",
    "NOTE: No adjacency matrix was used as input!\n",
    "\n",
    "==================================================================================================================\n",
    "\n",
    "SDePER settings are:\n",
    "\n",
    "* number of used CPU cores `n_core`: 64\n",
    "* **whether to use CVAE to remove platform effect `use_cvae`: false**\n",
    "\n",
    "ALL other options are left as default.\n",
    "\n",
    "Due to the absence of a command-line option to **disable additive platform effect term**, we manually adjusted the source code to disable it.\n",
    "\n",
    "==================================================================================================================\n",
    "\n",
    "the `bash` command to start cell type deconvolution is\n",
    "\n",
    "`runDeconvolution -q STARmap_cell_nUMI.csv -r scRNA_data_full.csv -c ref_scRNA_cell_celltype.csv -n 64 --use_cvae false`\n",
    "\n",
    "Note this Notebook uses **SDePER v1.5.0**. Cell type deconvolution result is renamed as [PlatEffDemo_ref_scRNA_SDePER_NO_PlatEffRmv_celltype_proportions.csv](https://github.com/az7jh2/SDePER_Analysis/blob/main/Simulation/Run_SDePER_on_simulation_data/PlatEffDemo/PlatEffDemo_ref_scRNA_SDePER_NO_PlatEffRmv_celltype_proportions.csv)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f43766d1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "SDePER (Spatial Deconvolution method with Platform Effect Removal) v1.5.0\n",
      "\n",
      "\n",
      "running options:\n",
      "spatial_file: /home/exouser/Spatial/STARmap_cell_nUMI.csv\n",
      "ref_file: /home/exouser/Spatial/scRNA_data_full.csv\n",
      "ref_celltype_file: /home/exouser/Spatial/ref_scRNA_cell_celltype.csv\n",
      "marker_file: None\n",
      "loc_file: None\n",
      "A_file: None\n",
      "n_cores: 64\n",
      "threshold: 0\n",
      "use_cvae: False\n",
      "use_imputation: False\n",
      "diagnosis: False\n",
      "verbose: True\n",
      "use_fdr: True\n",
      "p_val_cutoff: 0.05\n",
      "fc_cutoff: 1.2\n",
      "pct1_cutoff: 0.3\n",
      "pct2_cutoff: 0.1\n",
      "sortby_fc: True\n",
      "n_marker_per_cmp: 20\n",
      "filter_cell: True\n",
      "filter_gene: True\n",
      "n_hv_gene: 200\n",
      "n_pseudo_spot: 100000\n",
      "pseudo_spot_min_cell: 2\n",
      "pseudo_spot_max_cell: 8\n",
      "seq_depth_scaler: 10000\n",
      "cvae_input_scaler: 10\n",
      "cvae_init_lr: 0.01\n",
      "num_hidden_layer: 1\n",
      "use_batch_norm: True\n",
      "cvae_train_epoch: 500\n",
      "use_spatial_pseudo: False\n",
      "redo_de: True\n",
      "seed: 383\n",
      "lambda_r: [0.1, 0.268, 0.72, 1.931, 5.179, 13.895, 37.276, 100.0]\n",
      "lambda_g: [0.1, 0.268, 0.72, 1.931, 5.179, 13.895, 37.276, 100.0]\n",
      "diameter: 200\n",
      "impute_diameter: [160, 114, 80]\n",
      "\n",
      "\n",
      "######### Preprocessing... #########\n",
      "\n",
      "building CVAE skipped...\n",
      "\n",
      "read spatial data from file /home/exouser/Spatial/STARmap_cell_nUMI.csv\n",
      "total 2002 spots; 1020 genes\n",
      "\n",
      "filtering genes present in <3 spots: No genes removed\n",
      "\n",
      "filtering 0 mitochondrial genes\n",
      "\n",
      "finally remain 2002 spots; 1020 genes for downstream analysis\n",
      "\n",
      "no marker gene profile provided. Perform DE to get cell-type marker genes on scRNA-seq data...\n",
      "\n",
      "read scRNA-seq data from file /home/exouser/Spatial/scRNA_data_full.csv\n",
      "total 23178 cells; 45768 genes\n",
      "read scRNA-seq cell-type annotation from file /home/exouser/Spatial/ref_scRNA_cell_celltype.csv\n",
      "total 12 cell-types\n",
      "subset cells with cell-type annotation, finally keep 11835 cells; 45768 genes\n",
      "\n",
      "filtering cells with <200 genes: No cells removed\n",
      "\n",
      "filtering genes present in <10 cells: 12063 genes removed\n",
      "\n",
      "filtering 0 mitochondrial genes\n",
      "\n",
      "finally remain 11835 cells; 33705 genes for downstream analysis\n",
      "\n",
      "get 991 overlapped genes between spatial data and reference scRNA-seq data\n",
      "\n",
      "Differential analysis across cell-types on scRNA-seq data...\n",
      "0%...8%...17%...\n",
      "WARNING: only 13 genes passing filtering (<20) for Micro vs Endo\n",
      "\n",
      "WARNING: only 18 genes passing filtering (<20) for Micro vs PVALB\n",
      "\n",
      "WARNING: only 18 genes passing filtering (<20) for Micro vs SST\n",
      "\n",
      "WARNING: only 12 genes passing filtering (<20) for Micro vs Smc\n",
      "\n",
      "WARNING: only 19 genes passing filtering (<20) for Micro vs eL5\n",
      "\n",
      "WARNING: only 17 genes passing filtering (<20) for Micro vs eL6\n",
      "25%...\n",
      "WARNING: only 19 genes passing filtering (<20) for Oligo vs eL5\n",
      "\n",
      "WARNING: only 17 genes passing filtering (<20) for Oligo vs eL6\n",
      "33%...\n",
      "WARNING: only 11 genes passing filtering (<20) for PVALB vs SST\n",
      "42%...50%...58%...67%...\n",
      "WARNING: only 9 genes passing filtering (<20) for eL2/3 vs eL5\n",
      "\n",
      "WARNING: only 19 genes passing filtering (<20) for eL2/3 vs eL6\n",
      "75%...\n",
      "WARNING: only 10 genes passing filtering (<20) for eL4 vs eL2/3\n",
      "\n",
      "WARNING: only 2 genes passing filtering (<20) for eL4 vs eL5\n",
      "\n",
      "WARNING: only 19 genes passing filtering (<20) for eL4 vs eL6\n",
      "83%...\n",
      "WARNING: only 19 genes passing filtering (<20) for eL5 vs eL2/3\n",
      "\n",
      "WARNING: only 16 genes passing filtering (<20) for eL5 vs eL6\n",
      "92%...\n",
      "WARNING: only 8 genes passing filtering (<20) for eL6 vs eL5\n",
      "100%\n",
      "finally selected 466 cell-type marker genes\n",
      "\n",
      "\n",
      "gene filtering before modeling...\n",
      "all genes passed filtering\n",
      "\n",
      "spot filtering before modeling...\n",
      "all spots passed filtering\n",
      "\n",
      "\n",
      "######### Start GLRM modeling... #########\n",
      "\n",
      "GLRM settings:\n",
      "use SciPy minimize method:  L-BFGS-B\n",
      "global optimization turned off, local minimum will be used in GLRM\n",
      "use hybrid version of GLRM\n",
      "Numba detected total 64 available CPU cores. Use 64 CPU cores\n",
      "use 2001 points to calculate the heavy-tail density\n",
      "use weight threshold for Adaptive Lasso:  0.001\n",
      "total 60 unique nUMIs, min: 0.0, max: 120.0\n",
      "\n",
      "Build graph: \n",
      " Graph with 2002 nodes and 0 edges\n",
      "\n",
      "estimation of gene-specific platform effect gamma_g is skipped as already using CVAE to adjust platform effect\n",
      "\n",
      "\n",
      "Start GLRM fitting...\n",
      "\n",
      "first estimate MLE theta and corresponding e^alpha and sigma^2...\n",
      "\n",
      "GLRM model initialization...\n",
      "calculate MLE theta and sigma^2 iteratively...\n",
      "  iter | time_opt | time_sig | sigma2\n",
      "     0 |  149.986 |   37.335 |  1.433\n",
      "     1 |   90.575 |   18.822 |  1.572\n",
      "     2 |   82.438 |   18.542 |  1.614\n",
      "     3 |   71.348 |   14.384 |  1.627\n",
      "     4 |   53.604 |   14.370 |  1.630\n",
      "     5 |   33.789 |   14.576 |  1.631\n",
      "     6 |   15.738 |   11.001 |  1.632\n",
      "MLE theta and sigma^2 calculation finished. Elapsed time: 10.44 minutes.\n",
      "MLE theta estimation finished. Elapsed time: 10.44 minutes.\n",
      "\n",
      "calculate weights of Adaptive Lasso...\n",
      "\n",
      "Stage 1: variable selection using Adaptive Lasso starts with the MLE theta and e^alpha, using already estimated sigma^2 and gamma_g...\n",
      "specified hyper-parameter for Adaptive Lasso is: [0.1, 0.268, 0.72, 1.931, 5.179, 13.895, 37.276, 100.0]\n",
      "hyper-parameter for Adaptive Lasso: use cross-validation to find the optimal value from 8 candidates...\n",
      "\n",
      "Start cross-validation for hyper-parameter lambda_r...\n",
      "directly estimate theta by Adaptive Lasso loss function as NO Graph Laplacian constrain!\n",
      "0%...11%...22%...33%...44%...56%...67%...early stop\n",
      "find optimal lambda_r 0.720 with average negative log-likelihood 161914.7295 by 5 fold cross-validation. Elapsed time: 126.44 minutes.\n",
      "\n",
      "\n",
      "start ADMM iteration...\n",
      "  iter |  res_pri_n | res_dual_n |    eps_pri |   eps_dual |        rho |    new_rho | time_opt | time_reg | time_lap | tilde_RMSE |   hat_RMSE\n",
      "     0 |     14.796 |     14.796 |      0.234 |      0.234 |       1.00 |       1.00 |   13.717 |    0.002 |    0.009 |   0.133663 |   0.066831\n",
      "     1 |     14.714 |      0.202 |      0.234 |      0.249 |       1.00 |       2.00 |   60.242 |    0.001 |    0.004 |   0.132966 |   0.066483\n",
      "     2 |     14.564 |      0.981 |      0.234 |      0.278 |       2.00 |       4.00 |   61.734 |    0.001 |    0.004 |   0.131810 |   0.065905\n",
      "     3 |     13.675 |      8.305 |      0.233 |      0.332 |       4.00 |       8.00 |   68.349 |    0.000 |    0.004 |   0.124107 |   0.062053\n",
      "     4 |     11.731 |     28.139 |      0.247 |      0.423 |       8.00 |       8.00 |   76.221 |    0.001 |    0.004 |   0.105730 |   0.052865\n",
      "     5 |      9.899 |     35.671 |      0.255 |      0.494 |       8.00 |       8.00 |   71.950 |    0.001 |    0.003 |   0.088526 |   0.044263\n",
      "     6 |      8.771 |     31.719 |      0.251 |      0.550 |       8.00 |      16.00 |   77.520 |    0.001 |    0.004 |   0.078440 |   0.039220\n",
      "     7 |      7.545 |     38.891 |      0.258 |      0.650 |      16.00 |      16.00 |   75.245 |    0.001 |    0.004 |   0.067333 |   0.033666\n",
      "     8 |      6.421 |     43.025 |      0.262 |      0.733 |      16.00 |      16.00 |   81.616 |    0.001 |    0.004 |   0.057198 |   0.028599\n",
      "     9 |      5.650 |     38.117 |      0.257 |      0.800 |      16.00 |      32.00 |   75.146 |    0.001 |    0.004 |   0.050340 |   0.025170\n",
      "    10 |      4.820 |     48.153 |      0.267 |      0.920 |      32.00 |      32.00 |   77.359 |    0.001 |    0.004 |   0.042920 |   0.021460\n",
      "    11 |      4.066 |     54.553 |      0.274 |      1.020 |      32.00 |      32.00 |   70.653 |    0.001 |    0.004 |   0.036113 |   0.018056\n",
      "    12 |      3.537 |     49.667 |      0.269 |      1.098 |      32.00 |      64.00 |   67.380 |    0.001 |    0.004 |   0.031379 |   0.015690\n",
      "    13 |      2.983 |     63.385 |      0.283 |      1.238 |      64.00 |      64.00 |   70.264 |    0.001 |    0.004 |   0.026417 |   0.013208\n",
      "    14 |      2.505 |     69.685 |      0.289 |      1.353 |      64.00 |      64.00 |   66.054 |    0.001 |    0.004 |   0.022087 |   0.011043\n",
      "    15 |      2.181 |     62.828 |      0.282 |      1.443 |      64.00 |     128.00 |   64.075 |    0.001 |    0.004 |   0.019158 |   0.009579\n",
      "    16 |      1.821 |     78.608 |      0.298 |      1.604 |     128.00 |     128.00 |   61.371 |    0.001 |    0.004 |   0.015925 |   0.007963\n",
      "    17 |      1.514 |     86.651 |      0.306 |      1.735 |     128.00 |     128.00 |   63.314 |    0.001 |    0.004 |   0.013119 |   0.006559\n",
      "    18 |      1.312 |     76.893 |      0.296 |      1.838 |     128.00 |     256.00 |   63.290 |    0.001 |    0.004 |   0.011304 |   0.005652\n",
      "    19 |      1.086 |     96.008 |      0.315 |      2.018 |     256.00 |     256.00 |   61.623 |    0.001 |    0.004 |   0.009311 |   0.004655\n",
      "    20 |      0.899 |    104.193 |      0.323 |      2.164 |     256.00 |     256.00 |   55.478 |    0.001 |    0.007 |   0.007618 |   0.003809\n",
      "    21 |      0.782 |     91.532 |      0.311 |      2.279 |     256.00 |     512.00 |   53.445 |    0.001 |    0.004 |   0.006585 |   0.003293\n",
      "    22 |      0.649 |    111.128 |      0.330 |      2.482 |     512.00 |     512.00 |   52.580 |    0.001 |    0.005 |   0.005417 |   0.002708\n",
      "    23 |      0.538 |    119.985 |      0.339 |      2.651 |     512.00 |     512.00 |   50.624 |    0.001 |    0.004 |   0.004416 |   0.002208\n",
      "    24 |      0.469 |    104.233 |      0.323 |      2.785 |     512.00 |    1024.00 |   51.540 |    0.001 |    0.004 |   0.003805 |   0.001902\n",
      "    25 |      0.392 |    125.525 |      0.345 |      3.028 |    1024.00 |    1024.00 |   49.011 |    0.001 |    0.004 |   0.003147 |   0.001573\n",
      "    26 |      0.328 |    137.775 |      0.357 |      3.234 |    1024.00 |    1024.00 |   45.847 |    0.001 |    0.004 |   0.002577 |   0.001289\n",
      "    27 |      0.285 |    125.286 |      0.344 |      3.400 |    1024.00 |    2048.00 |   44.939 |    0.001 |    0.004 |   0.002199 |   0.001099\n",
      "    28 |      0.234 |    155.301 |      0.374 |      3.693 |    2048.00 |    2048.00 |   47.174 |    0.001 |    0.004 |   0.001773 |   0.000887\n",
      "    29 |      0.194 |    168.055 |      0.387 |      3.934 |    2048.00 |    2048.00 |   43.419 |    0.001 |    0.004 |   0.001415 |   0.000708\n",
      "    30 |      0.168 |    149.898 |      0.369 |      4.125 |    2048.00 |    4096.00 |   43.557 |    0.001 |    0.004 |   0.001199 |   0.000600\n",
      "    31 |      0.137 |    182.115 |      0.401 |      4.457 |    4096.00 |    4096.00 |   40.875 |    0.001 |    0.004 |   0.000947 |   0.000474\n",
      "    32 |      0.111 |    192.596 |      0.412 |      4.726 |    4096.00 |    4096.00 |   38.638 |    0.001 |    0.004 |   0.000732 |   0.000366\n",
      "    33 |      0.094 |    173.826 |      0.393 |      4.932 |    4096.00 |    8192.00 |   47.272 |    0.001 |    0.004 |   0.000597 |   0.000299\n",
      "    34 |      0.073 |    225.082 |      0.444 |      5.266 |    8192.00 |    8192.00 |   34.964 |    0.001 |    0.004 |   0.000443 |   0.000221\n",
      "    35 |      0.057 |    238.819 |      0.458 |      5.515 |    8192.00 |    8192.00 |   33.509 |    0.001 |    0.004 |   0.000321 |   0.000161\n",
      "    36 |      0.047 |    198.599 |      0.418 |      5.691 |    8192.00 |   16384.00 |   29.904 |    0.001 |    0.004 |   0.000256 |   0.000128\n",
      "    37 |      0.036 |    229.417 |      0.449 |      5.975 |   16384.00 |   16384.00 |   27.861 |    0.001 |    0.004 |   0.000183 |   0.000092\n",
      "    38 |      0.028 |    241.972 |      0.461 |      6.185 |   16384.00 |   16384.00 |   26.183 |    0.001 |    0.004 |   0.000127 |   0.000063\n",
      "    39 |      0.023 |    202.147 |      0.421 |      6.331 |   16384.00 |   32768.00 |   24.265 |    0.001 |    0.004 |   0.000098 |   0.000049\n",
      "    40 |      0.017 |    226.783 |      0.446 |      6.564 |   32768.00 |   32768.00 |   22.277 |    0.001 |    0.004 |   0.000067 |   0.000034\n",
      "    41 |      0.013 |    223.796 |      0.443 |      6.734 |   32768.00 |          / |   27.138 |    0.001 |    0.004 |   0.000045 |   0.000023\n",
      "early stop!\n",
      "Terminated (optimal) in 42 iterations.\n",
      "One optimization by ADMM finished. Elapsed time: 36.99 minutes.\n",
      "\n",
      "Stage 1 variable selection finished. Elapsed time: 163.43 minutes.\n",
      "\n",
      "Stage 2: final theta estimation with Graph Laplacian Constrain using already estimated sigma^2 and gamma_g\n",
      "HIGHLIGHT: reuse estimated theta and e^alpha in stage 1 as initial value\n",
      "No Adjacency Matrix of spots specified! Ignore Graph Laplacian Constrain in stage 2\n",
      "\n",
      "start ADMM iteration...\n",
      "  iter |  res_pri_n | res_dual_n |    eps_pri |   eps_dual |        rho |    new_rho | time_opt | time_reg | time_lap | tilde_RMSE |   hat_RMSE\n",
      "     0 |      2.136 |      2.136 |      0.221 |      0.221 |       1.00 |       1.00 |  111.910 |    0.001 |    0.004 |   0.017751 |   0.008875\n",
      "     1 |      0.002 |      2.136 |      0.221 |      0.221 |       1.00 |          / |   11.413 |    0.001 |    0.004 |   0.000004 |   0.000002\n",
      "early stop!\n",
      "Terminated (optimal) in 2 iterations.\n",
      "One optimization by ADMM finished. Elapsed time: 2.06 minutes.\n",
      "\n",
      "\n",
      "stage 2 finished. Elapsed time: 2.06 minutes.\n",
      "\n",
      "GLRM fitting finished. Elapsed time: 175.93 minutes.\n",
      "\n",
      "\n",
      "Post-processing estimated cell-type proportion theta...\n",
      "hard thresholding small theta values with threshold 0\n",
      "\n",
      "\n",
      "cell type deconvolution finished. Estimate results saved in /home/exouser/Spatial/celltype_proportions.csv. Elapsed time: 2.98 hours.\n",
      "\n",
      "\n",
      "######### No imputation #########\n",
      "\n",
      "\n",
      "whole pipeline finished. Total elapsed time: 2.98 hours.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "CompletedProcess(args='runDeconvolution -q STARmap_cell_nUMI.csv                           -r scRNA_data_full.csv                           -c ref_scRNA_cell_celltype.csv                           -n 64                           --use_cvae false\\n', returncode=0)"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import subprocess\n",
    "\n",
    "cmd = '''runDeconvolution -q STARmap_cell_nUMI.csv \\\n",
    "                          -r scRNA_data_full.csv \\\n",
    "                          -c ref_scRNA_cell_celltype.csv \\\n",
    "                          -n 64 \\\n",
    "                          --use_cvae false\n",
    "'''\n",
    "\n",
    "subprocess.run(cmd, check=True, text=True, shell=True)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
