---
title: "Generate *downsampled* scRNA-seq reference data for simulation analysis"
author: "Ningshan Li"
date: "2024/05/21"
output: 
  html_notebook:
    code_folding: hide
    highlight: tango
    number_sections: yes
    theme: united
    toc: yes
    toc_depth: 6
    toc_float: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, eval = TRUE, warning = FALSE, message = FALSE, results='hold', fig.width = 7, fig.height = 5, dpi = 300)


library(dplyr)
library(ggplot2)


`%notin%` = Negate(`%in%`)

set.seed(1)

home.dir = '/home/hill103/Documents/Spatial/Simulation/Generate_downsampled_ref_data'


my.distinct.colors20 = c("#e6194b", "#3cb44b", "#ffe119", "#4363d8", "#f58231", "#911eb4", "#46f0f0", "#f032e6", "#bcf60c", "#fabebe", "#008080", "#9a6324", "#800000", "#aaffc3", "#808000", "#000075", "#808080", "#e6beff", "#ffd8b1", "#000000")

my.distinct.colors40 = c("#00ff00","#ff4500","#00ced1","#556b2f","#a0522d","#8b0000","#808000","#483d8b","#008000","#008080","#4682b4","#000080","#9acd32","#daa520","#7f007f","#8fbc8f","#b03060","#d2b48c","#696969","#ff8c00","#00ff7f","#dc143c","#f4a460","#0000ff","#a020f0","#adff2f","#ff00ff","#1e90ff","#f0e68c","#fa8072","#ffff54","#dda0dd","#87ceeb","#7b68ee","#ee82ee","#98fb98","#7fffd4","#ffb6c1","#dcdcdc","#000000")
```



# Summary

We generated internal and external scRNA-seq reference data by downsampling the number of cells of cell type **Astro** and **Oligo**. We randomly downsampled the number of cells to 5, 10, 20, 50, 100 and 250.

A smaller number of cells per cell type hinders the generation of a robust cell type-specific gene expression profile, and also affects the selection of cell type marker genes in differential analysis.



# Version

```{r}
version[['version.string']]
```


# Read cell type annotation

## Internal reference -- STARmap cells

We read cell type annotation of all 2,002 STARmap cells (available at [STARmap_cell_celltype.csv](https://github.com/az7jh2/SDePER_Analysis/blob/main/Simulation/Run_SDePER_on_simulation_data/Scenario_1/ref_spatial/STARmap_cell_celltype.csv)), and randomly downsampled the number of Astro and Oligo cells.


```{r}
file_name = file.path(home.dir, 'STARmap_cell_celltype.csv')
int_df = read.csv(file_name, row.names = 1, check.names = F, stringsAsFactors = F)
print(sprintf('load data from %s', file_name))
print(sprintf('total %d cells', nrow(int_df)))
```

Number of cells of each cell type is shown as below.

```{r}
table(int_df$celltype)
```


## External reference -- [GSE115746](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE115746) scRNA-seq cells

We read cell type annotation of all 11,835 scRNA-seq cells (available at [ref_scRNA_cell_celltype.csv](https://github.com/az7jh2/SDePER_Analysis/blob/main/Simulation/Run_SDePER_on_simulation_data/Scenario_1/ref_scRNA_seq/ref_scRNA_cell_celltype.csv)), and randomly downsampled the number of Astro and Oligo cells.

```{r}
file_name = file.path(home.dir, 'ref_scRNA_cell_celltype.csv')
ext_df = read.csv(file_name, row.names = 1, check.names = F, stringsAsFactors = F)
print(sprintf('load data from %s', file_name))
print(sprintf('total %d cells', nrow(ext_df)))
```

Number of cells of each cell type is shown as below.

```{r}
table(ext_df$celltype)
```



# Downsampling Astro cells

## Internal reference

Randomly selected a specific number of cells to ensure there are 5, 10, 20, 50, 100 and 250 cells for Astro, respectively.

```{r}
target_celltype = 'Astro'
downsample_ref = list()
total_num = sum(int_df$celltype == target_celltype)

for (num in c(5, 10, 20, 50, 100, 250)) {
  unselect_num = total_num - num
  unselect_cells = sample(row.names(int_df[int_df$celltype==target_celltype, , drop=F]), unselect_num, replace = F)
  downsample_ref[[paste0(num, 'cells')]] = int_df[!row.names(int_df) %in% unselect_cells, , drop=F]
}
```


Double check.

```{r}
for (one_setting in names(downsample_ref)) {
  print(one_setting)
  print(table(downsample_ref[[one_setting]]$celltype))
}
```


Save result. The downsampled cell type annotation files are available in [this folder](https://github.com/az7jh2/SDePER_Analysis/tree/main/Simulation/Run_SDePER_on_simulation_data_with_downsampled_ref/Astro/Scenario_1/ref_spatial).

```{r}
for (one_setting in names(downsample_ref)) {
  file_name = paste0('STARmap_cell_annotation_', target_celltype, '_', one_setting, '.csv')
  write.csv(downsample_ref[[one_setting]], file_name)
  print(sprintf('save cell type annotation to file %s', file_name))
}
```


## External reference

Randomly selected a specific number of cells to ensure there are 5, 10, 20, 50, 100 and 250 cells for Astro, respectively.

```{r}
target_celltype = 'Astro'
downsample_ref = list()
total_num = sum(ext_df$celltype == target_celltype)

for (num in c(5, 10, 20, 50, 100, 250)) {
  unselect_num = total_num - num
  unselect_cells = sample(row.names(ext_df[ext_df$celltype==target_celltype, , drop=F]), unselect_num, replace = F)
  downsample_ref[[paste0(num, 'cells')]] = ext_df[!row.names(ext_df) %in% unselect_cells, , drop=F]
}
```


Double check.

```{r}
for (one_setting in names(downsample_ref)) {
  print(one_setting)
  print(table(downsample_ref[[one_setting]]$celltype))
}
```


Save result. The downsampled cell type annotation files are available in [this folder](https://github.com/az7jh2/SDePER_Analysis/tree/main/Simulation/Run_SDePER_on_simulation_data_with_downsampled_ref/Astro/Scenario_1/ref_scRNA_seq).

```{r}
for (one_setting in names(downsample_ref)) {
  file_name = paste0('scRNA_cell_annotation_', target_celltype, '_', one_setting, '.csv')
  write.csv(downsample_ref[[one_setting]], file_name)
  print(sprintf('save cell type annotation to file %s', file_name))
}
```



# Downsampling Oligo cells

## Internal reference

Randomly selected a specific number of cells to ensure there are 5, 10, 20 and 50 cells for Oligo, respectively.

```{r}
target_celltype = 'Oligo'
downsample_ref = list()
total_num = sum(int_df$celltype == target_celltype)

for (num in c(5, 10, 20, 50)) {
  unselect_num = total_num - num
  unselect_cells = sample(row.names(int_df[int_df$celltype==target_celltype, , drop=F]), unselect_num, replace = F)
  downsample_ref[[paste0(num, 'cells')]] = int_df[!row.names(int_df) %in% unselect_cells, , drop=F]
}
```


Double check.

```{r}
for (one_setting in names(downsample_ref)) {
  print(one_setting)
  print(table(downsample_ref[[one_setting]]$celltype))
}
```


Save result. The downsampled cell type annotation files are available in [this folder](https://github.com/az7jh2/SDePER_Analysis/tree/main/Simulation/Run_SDePER_on_simulation_data_with_downsampled_ref/Oligo/Scenario_1/ref_spatial).

```{r}
for (one_setting in names(downsample_ref)) {
  file_name = paste0('STARmap_cell_annotation_', target_celltype, '_', one_setting, '.csv')
  write.csv(downsample_ref[[one_setting]], file_name)
  print(sprintf('save cell type annotation to file %s', file_name))
}
```


## External reference

Randomly selected a specific number of cells to ensure there are 5, 10, 20 and 50 cells for Oligo, respectively.

```{r}
target_celltype = 'Oligo'
downsample_ref = list()
total_num = sum(ext_df$celltype == target_celltype)

for (num in c(5, 10, 20, 50)) {
  unselect_num = total_num - num
  unselect_cells = sample(row.names(ext_df[ext_df$celltype==target_celltype, , drop=F]), unselect_num, replace = F)
  downsample_ref[[paste0(num, 'cells')]] = ext_df[!row.names(ext_df) %in% unselect_cells, , drop=F]
}
```


Double check.

```{r}
for (one_setting in names(downsample_ref)) {
  print(one_setting)
  print(table(downsample_ref[[one_setting]]$celltype))
}
```


Save result. The downsampled cell type annotation files are available in [this folder](https://github.com/az7jh2/SDePER_Analysis/tree/main/Simulation/Run_SDePER_on_simulation_data_with_downsampled_ref/Oligo/Scenario_1/ref_scRNA_seq).

```{r}
for (one_setting in names(downsample_ref)) {
  file_name = paste0('scRNA_cell_annotation_', target_celltype, '_', one_setting, '.csv')
  write.csv(downsample_ref[[one_setting]], file_name)
  print(sprintf('save cell type annotation to file %s', file_name))
}
```


