{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "db496b92",
   "metadata": {},
   "source": [
    "# Run *SDePER* on sequencing-based simulated data: Scenario 1 + scRNA-seq data as reference + NO platform effect removal\n",
    "\n",
    "In this Notebook we run ablation test on SDePER on simulated data. For generating **sequencing-based** simulated data via coarse-graining procedure please refer [generate_simulated_spatial_data.nb.html](https://rawcdn.githack.com/az7jh2/SDePER_Analysis/c963d08f74f4591c2ef6f132177795297793d878/Simulation_seq_based/Generate_simulation_data/generate_simulated_spatial_data.nb.html) in [Simulation_seq_based](https://github.com/az7jh2/SDePER_Analysis/tree/main/Simulation_seq_based) folder.\n",
    "\n",
    "**Scenario 1** means the reference data for deconvolution includes all single cells with the **matched 12 cell types**.\n",
    "\n",
    "**scRNA-seq data as reference** means the reference data is another scRNA-seq data ([GSE115746](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE115746)) from the same tissue with simulated spatial data, therefore **platform effect exists**.\n",
    "\n",
    "**NO platform effect removal** means we conduct cell type deconvolution **disregarding platform effect**, meaning that neither CVAE nor an additive gene-wise platform effect term is utilized.\n",
    "\n",
    "==================================================================================================================\n",
    "\n",
    "So here we use the **4 input files** as shown below:\n",
    "\n",
    "1. raw nUMI counts of simulated spatial transcriptomic data (spots × genes): [sim_seq_based_spatial_spot_nUMI.csv](https://github.com/az7jh2/SDePER_Analysis/blob/main/Simulation_seq_based/Generate_simulation_data/sim_seq_based_spatial_spot_nUMI.csv)\n",
    "2. raw nUMI counts of reference scRNA-seq data (cells × genes): `scRNA_data_full.csv`. Since the file size of csv file of raw nUMI matrix of all 23,178 cells and 45,768 genes is up to 2.3 GB, we do not provide this file in our repository. It's just a **matrix transpose** of [GSE115746_cells_exon_counts.csv.gz](https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE115746&format=file&file=GSE115746%5Fcells%5Fexon%5Fcounts%2Ecsv%2Egz) in [GSE115746](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE115746) to satisty the file format requirement that rows as cells and columns as genes.\n",
    "3. cell type annotations for cells of **the matched 12 cell types** in reference scRNA-seq data (cells × 1): [ref_scRNA_cell_celltype.csv](https://github.com/az7jh2/SDePER_Analysis/blob/main/Simulation/Run_SDePER_on_simulation_data/Scenario_1/ref_scRNA_seq/ref_scRNA_cell_celltype.csv)\n",
    "4. adjacency matrix of spots in simulated spatial transcriptomic data (spots × spots): [sim_spatial_spot_adjacency_matrix.csv](https://github.com/az7jh2/SDePER_Analysis/blob/main/Simulation/Generate_simulation_data/sim_spatial_spot_adjacency_matrix.csv)\n",
    "\n",
    "==================================================================================================================\n",
    "\n",
    "SDePER settings are the same as baseline run [S1_ref_scRNA_SDePER_WITH_CVAE.ipynb](https://github.com/az7jh2/SDePER_Analysis/blob/main/Simulation_seq_based/Run_SDePER_on_simulation_data/Scenario_1/ref_scRNA_seq/S1_ref_scRNA_SDePER_WITH_CVAE.ipynb), and we discarded unneeded command-line options:\n",
    "\n",
    "* number of selected TOP marker genes for each comparison in Differential `n_marker_per_cmp`: 20\n",
    "* number of used CPU cores `n_core`: 64\n",
    "\n",
    "ALL other options are left as default.\n",
    "\n",
    "**For ablation test, disable CVAE and additive platform effect term**.\n",
    "\n",
    "Due to the absence of a command-line option to disable additive platform effect term, we manually adjusted the source code to disable it.\n",
    "\n",
    "==================================================================================================================\n",
    "\n",
    "the `bash` command to start cell type deconvolution is\n",
    "\n",
    "`runDeconvolution -q sim_seq_based_spatial_spot_nUMI.csv -r scRNA_data_full.csv -c ref_scRNA_cell_celltype.csv -a sim_spatial_spot_adjacency_matrix.csv --n_marker_per_cmp 20 -n 64 --use_cvae false`\n",
    "\n",
    "Note this Notebook uses **SDePER v1.2.1**. Cell type deconvolution result is renamed as [S1_ref_scRNA_SDePER_Ablation_NO_PlatEffRmv_celltype_proportions.csv](https://github.com/az7jh2/SDePER_Analysis/blob/main/Ablation/Ablation_simulation_seq_based/S1_ref_scRNA_SDePER_Ablation_NO_PlatEffRmv_celltype_proportions.csv)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "df9e3939",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "SDePER (Spatial Deconvolution method with Platform Effect Removal) v1.2.1\n",
      "\n",
      "\n",
      "running options:\n",
      "spatial_file: /home/exouser/Spatial/sim_seq_based_spatial_spot_nUMI.csv\n",
      "ref_file: /home/exouser/Spatial/scRNA_data_full.csv\n",
      "ref_celltype_file: /home/exouser/Spatial/ref_scRNA_cell_celltype.csv\n",
      "marker_file: None\n",
      "loc_file: None\n",
      "A_file: /home/exouser/Spatial/sim_spatial_spot_adjacency_matrix.csv\n",
      "n_cores: 64\n",
      "threshold: 0\n",
      "use_cvae: False\n",
      "use_imputation: False\n",
      "diagnosis: False\n",
      "verbose: True\n",
      "use_fdr: True\n",
      "p_val_cutoff: 0.05\n",
      "fc_cutoff: 1.2\n",
      "pct1_cutoff: 0.3\n",
      "pct2_cutoff: 0.1\n",
      "sortby_fc: True\n",
      "n_marker_per_cmp: 20\n",
      "filter_cell: True\n",
      "filter_gene: True\n",
      "n_hv_gene: 200\n",
      "n_pseudo_spot: 500000\n",
      "pseudo_spot_min_cell: 2\n",
      "pseudo_spot_max_cell: 8\n",
      "seq_depth_scaler: 10000\n",
      "cvae_input_scaler: 10\n",
      "cvae_init_lr: 0.01\n",
      "num_hidden_layer: 2\n",
      "use_batch_norm: True\n",
      "cvae_train_epoch: 500\n",
      "use_spatial_pseudo: False\n",
      "redo_de: True\n",
      "seed: 383\n",
      "lambda_r: [0.1, 0.268, 0.72, 1.931, 5.179, 13.895, 37.276, 100.0]\n",
      "lambda_g: [0.1, 0.268, 0.72, 1.931, 5.179, 13.895, 37.276, 100.0]\n",
      "diameter: 200\n",
      "impute_diameter: [160, 114, 80]\n",
      "\n",
      "\n",
      "######### Preprocessing... #########\n",
      "\n",
      "building CVAE skipped...\n",
      "\n",
      "read spatial data from file /home/exouser/Spatial/sim_seq_based_spatial_spot_nUMI.csv\n",
      "total 581 spots; 25187 genes\n",
      "\n",
      "filtering genes present in <3 spots: 9510 genes removed\n",
      "\n",
      "no marker gene profile provided. Perform DE to get cell-type marker genes on scRNA-seq data...\n",
      "\n",
      "read scRNA-seq data from file /home/exouser/Spatial/scRNA_data_full.csv\n",
      "total 23178 cells; 45768 genes\n",
      "read scRNA-seq cell-type annotation from file /home/exouser/Spatial/ref_scRNA_cell_celltype.csv\n",
      "total 12 cell-types\n",
      "subset cells with cell-type annotation, finally keep 11835 cells; 45768 genes\n",
      "\n",
      "filtering cells with <200 genes: No cells removed\n",
      "\n",
      "filtering genes present in <10 cells: 12063 genes removed\n",
      "\n",
      "Differential analysis across cell-types on scRNA-seq data...\n",
      "0%...8%...17%...25%...33%...42%...50%...58%...67%...75%...83%...92%...finally selected 748 cell-type marker genes\n",
      "\n",
      "\n",
      "gene filtering before modeling...\n",
      "8 genes with nUMIs<5 in all spatial spots and need to be excluded\n",
      "finally use 740 genes for modeling\n",
      "\n",
      "spot filtering before modeling...\n",
      "all spots passed filtering\n",
      "\n",
      "\n",
      "######### Start GLRM modeling... #########\n",
      "\n",
      "GLRM settings:\n",
      "use SciPy minimize method:  L-BFGS-B\n",
      "global optimization turned off, local minimum will be used in GLRM\n",
      "use hybrid version of GLRM\n",
      "Numba detected total 64 available CPU cores. Use 64 CPU cores\n",
      "use 2001 points to calculate the heavy-tail density\n",
      "use weight threshold for Adaptive Lasso:  0.001\n",
      "total 44 unique nUMIs, min: 0, max: 142\n",
      "\n",
      "Build graph: \n",
      " Graph with 581 nodes and 1029 edges\n",
      "\n",
      "estimation of gene-specific platform effect gamma_g is skipped as already using CVAE to adjust platform effect\n",
      "\n",
      "\n",
      "Start GLRM fitting...\n",
      "\n",
      "first estimate MLE theta and corresponding e^alpha and sigma^2...\n",
      "\n",
      "GLRM model initialization...\n",
      "calculate MLE theta and sigma^2 iteratively...\n",
      "  iter | time_opt | time_sig | sigma2\n",
      "     0 |   35.637 |    5.085 |  0.025\n",
      "     1 |   27.314 |    1.009 |  0.010\n",
      "     2 |   13.019 |    0.503 |  0.010\n",
      "MLE theta and sigma^2 calculation finished. Elapsed time: 1.38 minutes.\n",
      "MLE theta estimation finished. Elapsed time: 1.38 minutes.\n",
      "\n",
      "calculate weights of Adaptive Lasso...\n",
      "\n",
      "Stage 1: variable selection using Adaptive Lasso starts with the MLE theta and e^alpha, using already estimated sigma^2 and gamma_g...\n",
      "specified hyper-parameter for Adaptive Lasso is: [0.1, 0.268, 0.72, 1.931, 5.179, 13.895, 37.276, 100.0]\n",
      "hyper-parameter for Adaptive Lasso: use cross-validation to find the optimal value from 8 candidates...\n",
      "\n",
      "Start cross-validation for hyper-parameter lambda_r...\n",
      "directly estimate theta by Adaptive Lasso loss function as NO Graph Laplacian constrain!\n",
      "0%...11%...22%...33%...44%...56%...67%...early stop\n",
      "find optimal lambda_r 0.720 with average negative log-likelihood 54013.0185 by 5 fold cross-validation. Elapsed time: 13.67 minutes.\n",
      "\n",
      "\n",
      "start ADMM iteration...\n",
      "  iter |  res_pri_n | res_dual_n |    eps_pri |   eps_dual |        rho |    new_rho | time_opt | time_reg | time_lap | tilde_RMSE |   hat_RMSE\n",
      "     0 |     10.870 |     10.871 |      0.129 |      0.129 |       1.00 |       1.00 |    2.928 |    0.000 |    0.004 |   0.181239 |   0.090620\n",
      "     1 |     10.833 |      0.078 |      0.129 |      0.140 |       1.00 |       2.00 |   13.091 |    0.000 |    0.003 |   0.180640 |   0.090320\n",
      "     2 |      9.165 |      5.807 |      0.127 |      0.158 |       2.00 |       2.00 |   12.952 |    0.000 |    0.003 |   0.153464 |   0.076732\n",
      "     3 |      7.107 |      8.120 |      0.126 |      0.170 |       2.00 |       4.00 |   10.824 |    0.000 |    0.003 |   0.109743 |   0.054871\n",
      "     4 |      5.706 |     11.168 |      0.129 |      0.189 |       4.00 |       4.00 |   13.211 |    0.000 |    0.003 |   0.088835 |   0.044418\n",
      "     5 |      4.718 |     10.777 |      0.129 |      0.202 |       4.00 |       4.00 |   12.398 |    0.000 |    0.003 |   0.074692 |   0.037346\n",
      "     6 |      4.275 |      8.004 |      0.126 |      0.212 |       4.00 |       8.00 |   12.618 |    0.000 |    0.003 |   0.067126 |   0.033563\n",
      "     7 |      3.704 |      9.160 |      0.127 |      0.233 |       8.00 |       8.00 |   14.426 |    0.000 |    0.003 |   0.057277 |   0.028639\n",
      "     8 |      3.153 |     10.710 |      0.129 |      0.252 |       8.00 |       8.00 |   14.116 |    0.000 |    0.003 |   0.048462 |   0.024231\n",
      "     9 |      2.844 |      9.218 |      0.127 |      0.269 |       8.00 |      16.00 |   12.798 |    0.000 |    0.003 |   0.043923 |   0.021961\n",
      "    10 |      2.486 |     11.993 |      0.130 |      0.300 |      16.00 |      16.00 |   14.164 |    0.000 |    0.003 |   0.038513 |   0.019257\n",
      "    11 |      2.140 |     14.036 |      0.132 |      0.326 |      16.00 |      16.00 |   13.702 |    0.000 |    0.003 |   0.033011 |   0.016506\n",
      "    12 |      1.922 |     12.130 |      0.130 |      0.348 |      16.00 |      32.00 |   13.384 |    0.000 |    0.003 |   0.029703 |   0.014852\n",
      "    13 |      1.669 |     15.879 |      0.134 |      0.387 |      32.00 |      32.00 |   14.604 |    0.000 |    0.003 |   0.025736 |   0.012868\n",
      "    14 |      1.436 |     18.037 |      0.136 |      0.421 |      32.00 |      32.00 |   14.140 |    0.000 |    0.003 |   0.021967 |   0.010983\n",
      "    15 |      1.281 |     15.869 |      0.134 |      0.450 |      32.00 |      64.00 |   13.654 |    0.000 |    0.003 |   0.019658 |   0.009829\n",
      "    16 |      1.101 |     21.337 |      0.139 |      0.502 |      64.00 |      64.00 |   14.434 |    0.000 |    0.003 |   0.017008 |   0.008504\n",
      "    17 |      0.939 |     24.162 |      0.142 |      0.546 |      64.00 |      64.00 |   13.658 |    0.000 |    0.003 |   0.014538 |   0.007269\n",
      "    18 |      0.835 |     21.005 |      0.139 |      0.582 |      64.00 |     128.00 |   13.509 |    0.000 |    0.003 |   0.012984 |   0.006492\n",
      "    19 |      0.720 |     26.965 |      0.145 |      0.649 |     128.00 |     128.00 |   13.272 |    0.000 |    0.003 |   0.011237 |   0.005619\n",
      "    20 |      0.624 |     30.377 |      0.148 |      0.707 |     128.00 |     128.00 |   12.747 |    0.000 |    0.003 |   0.009719 |   0.004860\n",
      "    21 |      0.554 |     27.637 |      0.146 |      0.756 |     128.00 |     256.00 |   12.819 |    0.000 |    0.003 |   0.008616 |   0.004308\n",
      "    22 |      0.481 |     36.406 |      0.154 |      0.845 |     256.00 |     256.00 |   12.509 |    0.000 |    0.003 |   0.007462 |   0.003731\n",
      "    23 |      0.419 |     40.643 |      0.159 |      0.924 |     256.00 |     256.00 |   15.383 |    0.000 |    0.003 |   0.006482 |   0.003241\n",
      "    24 |      0.377 |     37.170 |      0.155 |      0.991 |     256.00 |     512.00 |   11.411 |    0.000 |    0.003 |   0.005839 |   0.002919\n",
      "    25 |      0.333 |     46.858 |      0.165 |      1.118 |     512.00 |     512.00 |   10.997 |    0.000 |    0.003 |   0.005147 |   0.002574\n",
      "    26 |      0.292 |     53.144 |      0.171 |      1.230 |     512.00 |     512.00 |   10.667 |    0.000 |    0.003 |   0.004503 |   0.002251\n",
      "    27 |      0.262 |     52.015 |      0.170 |      1.326 |     512.00 |    1024.00 |   10.443 |    0.000 |    0.003 |   0.004018 |   0.002009\n",
      "    28 |      0.231 |     67.647 |      0.186 |      1.504 |    1024.00 |    1024.00 |   10.415 |    0.000 |    0.003 |   0.003541 |   0.001771\n",
      "    29 |      0.203 |     76.792 |      0.195 |      1.660 |    1024.00 |    1024.00 |   10.016 |    0.000 |    0.003 |   0.003096 |   0.001548\n",
      "    30 |      0.182 |     73.635 |      0.192 |      1.794 |    1024.00 |    2048.00 |   10.124 |    0.000 |    0.003 |   0.002766 |   0.001383\n",
      "    31 |      0.156 |     99.842 |      0.218 |      2.032 |    2048.00 |    2048.00 |   10.130 |    0.000 |    0.003 |   0.002377 |   0.001188\n",
      "    32 |      0.134 |    112.995 |      0.231 |      2.231 |    2048.00 |    2048.00 |    9.569 |    0.000 |    0.003 |   0.002015 |   0.001008\n",
      "    33 |      0.119 |    102.920 |      0.221 |      2.395 |    2048.00 |    4096.00 |   12.613 |    0.000 |    0.003 |   0.001782 |   0.000891\n",
      "    34 |      0.102 |    129.771 |      0.248 |      2.686 |    4096.00 |    4096.00 |    9.455 |    0.000 |    0.003 |   0.001519 |   0.000760\n",
      "    35 |      0.088 |    140.375 |      0.258 |      2.938 |    4096.00 |    4096.00 |    8.749 |    0.000 |    0.003 |   0.001290 |   0.000645\n",
      "    36 |      0.079 |    126.761 |      0.245 |      3.153 |    4096.00 |    8192.00 |    8.472 |    0.000 |    0.003 |   0.001142 |   0.000571\n",
      "    37 |      0.067 |    161.506 |      0.280 |      3.539 |    8192.00 |    8192.00 |    8.738 |    0.000 |    0.003 |   0.000957 |   0.000478\n",
      "    38 |      0.057 |    182.415 |      0.300 |      3.866 |    8192.00 |    8192.00 |    8.023 |    0.000 |    0.003 |   0.000800 |   0.000400\n",
      "    39 |      0.051 |    162.559 |      0.281 |      4.138 |    8192.00 |   16384.00 |    7.856 |    0.000 |    0.003 |   0.000702 |   0.000351\n",
      "    40 |      0.043 |    204.446 |      0.323 |      4.620 |   16384.00 |   16384.00 |    7.497 |    0.000 |    0.003 |   0.000580 |   0.000290\n",
      "    41 |      0.036 |    227.710 |      0.346 |      5.019 |   16384.00 |   16384.00 |    7.477 |    0.000 |    0.003 |   0.000472 |   0.000236\n",
      "    42 |      0.030 |    211.160 |      0.329 |      5.335 |   16384.00 |   32768.00 |    6.878 |    0.000 |    0.003 |   0.000397 |   0.000199\n",
      "    43 |      0.025 |    262.255 |      0.380 |      5.868 |   32768.00 |   32768.00 |    6.753 |    0.000 |    0.003 |   0.000314 |   0.000157\n",
      "    44 |      0.020 |    281.890 |      0.400 |      6.288 |   32768.00 |   32768.00 |    6.312 |    0.000 |    0.003 |   0.000243 |   0.000122\n",
      "    45 |      0.017 |    250.338 |      0.368 |      6.603 |   32768.00 |   65536.00 |    6.234 |    0.000 |    0.003 |   0.000198 |   0.000099\n",
      "    46 |      0.014 |    298.646 |      0.417 |      7.134 |   65536.00 |   65536.00 |    5.622 |    0.000 |    0.003 |   0.000152 |   0.000076\n",
      "    47 |      0.011 |    308.517 |      0.427 |      7.556 |   65536.00 |   65536.00 |    5.600 |    0.000 |    0.003 |   0.000114 |   0.000057\n",
      "    48 |      0.009 |    276.975 |      0.395 |      7.875 |   65536.00 |  131072.00 |    5.227 |    0.000 |    0.003 |   0.000091 |   0.000045\n",
      "    49 |      0.007 |    357.455 |      0.476 |      8.370 |  131072.00 |          / |    4.881 |    0.000 |    0.003 |   0.000064 |   0.000032\n",
      "early stop!\n",
      "Terminated (optimal) in 50 iterations.\n",
      "One optimization by ADMM finished. Elapsed time: 8.88 minutes.\n",
      "\n",
      "Stage 1 variable selection finished. Elapsed time: 22.54 minutes.\n",
      "\n",
      "Stage 2: final theta estimation with Graph Laplacian Constrain using already estimated sigma^2 and gamma_g\n",
      "specified hyper-parameter for Graph Laplacian Constrain is: [0.1, 0.268, 0.72, 1.931, 5.179, 13.895, 37.276, 100.0]\n",
      "hyper-parameter for Graph Laplacian Constrain: use cross-validation to find the optimal value from 8 candidates...\n",
      "\n",
      "Start cross-validation for hyper-parameter lambda_g...\n",
      "still use ADMM even NO Graph Laplacian constrain (lambda_g=0)\n",
      "0%...11%...22%...33%...44%...56%...67%...78%...89%...100%\n",
      "find optimal lambda_g 13.895 with average negative log-likelihood 53984.7757 by 5 fold cross-validation. Elapsed time: 115.14 minutes.\n",
      "\n",
      "\n",
      "start ADMM iteration...\n",
      "  iter |  res_pri_n | res_dual_n |    eps_pri |   eps_dual |        rho |    new_rho | time_opt | time_reg | time_lap | tilde_RMSE |   hat_RMSE\n",
      "     0 |     12.230 |      9.922 |      0.130 |      0.130 |       1.00 |       1.00 |   21.826 |    0.000 |    0.007 |   0.139381 |   0.120390\n",
      "     1 |      9.826 |      7.167 |      0.128 |      0.138 |       1.00 |       1.00 |    7.659 |    0.000 |    0.007 |   0.120367 |   0.100445\n",
      "     2 |      8.668 |      7.444 |      0.127 |      0.146 |       1.00 |       1.00 |    7.754 |    0.000 |    0.007 |   0.095171 |   0.090071\n",
      "     3 |      8.131 |      7.730 |      0.126 |      0.154 |       1.00 |       2.00 |    7.710 |    0.000 |    0.007 |   0.087309 |   0.084746\n",
      "     4 |      7.639 |     12.918 |      0.131 |      0.168 |       2.00 |       2.00 |    9.507 |    0.000 |    0.007 |   0.082592 |   0.079602\n",
      "     5 |      7.053 |     13.260 |      0.131 |      0.182 |       2.00 |       2.00 |    9.154 |    0.000 |    0.007 |   0.076044 |   0.073787\n",
      "     6 |      6.518 |     14.066 |      0.132 |      0.195 |       2.00 |       4.00 |    8.832 |    0.000 |    0.007 |   0.068704 |   0.068672\n",
      "     7 |      6.065 |     23.077 |      0.141 |      0.219 |       4.00 |       4.00 |   10.456 |    0.000 |    0.007 |   0.065633 |   0.063206\n",
      "     8 |      5.445 |     23.934 |      0.142 |      0.240 |       4.00 |       4.00 |    9.864 |    0.000 |    0.007 |   0.059212 |   0.056554\n",
      "     9 |      4.829 |     25.473 |      0.144 |      0.259 |       4.00 |       8.00 |   10.112 |    0.000 |    0.007 |   0.050328 |   0.050535\n",
      "    10 |      4.299 |     39.997 |      0.158 |      0.292 |       8.00 |       8.00 |   11.057 |    0.000 |    0.007 |   0.046819 |   0.044352\n",
      "    11 |      3.629 |     41.387 |      0.159 |      0.320 |       8.00 |       8.00 |   10.422 |    0.000 |    0.007 |   0.039700 |   0.037280\n",
      "    12 |      3.005 |     43.521 |      0.162 |      0.343 |       8.00 |      16.00 |   10.009 |    0.000 |    0.007 |   0.030668 |   0.031127\n",
      "    13 |      2.514 |     64.022 |      0.182 |      0.381 |      16.00 |      16.00 |   10.976 |    0.000 |    0.007 |   0.027426 |   0.025579\n",
      "    14 |      1.926 |     65.622 |      0.184 |      0.409 |      16.00 |      16.00 |   10.305 |    0.000 |    0.007 |   0.021440 |   0.019489\n",
      "    15 |      1.415 |     67.840 |      0.186 |      0.430 |      16.00 |      32.00 |    9.256 |    0.000 |    0.007 |   0.013888 |   0.014493\n",
      "    16 |      1.102 |     91.530 |      0.210 |      0.461 |      32.00 |      32.00 |   10.022 |    0.000 |    0.006 |   0.011660 |   0.011033\n",
      "    17 |      0.737 |     92.762 |      0.211 |      0.481 |      32.00 |      32.00 |   10.709 |    0.000 |    0.006 |   0.008614 |   0.007265\n",
      "    18 |      0.437 |     94.375 |      0.212 |      0.493 |      32.00 |      64.00 |    7.790 |    0.000 |    0.006 |   0.003967 |   0.004380\n",
      "    19 |      0.337 |    115.918 |      0.234 |      0.508 |      64.00 |      64.00 |    8.321 |    0.000 |    0.005 |   0.002890 |   0.003253\n",
      "    20 |      0.203 |    116.509 |      0.235 |      0.516 |      64.00 |      64.00 |    8.058 |    0.000 |    0.005 |   0.002631 |   0.001820\n",
      "    21 |      0.089 |    117.409 |      0.235 |      0.518 |      64.00 |     128.00 |   10.266 |    0.000 |    0.006 |   0.000865 |   0.000764\n",
      "    22 |      0.093 |    133.503 |      0.252 |      0.522 |     128.00 |     128.00 |    5.588 |    0.000 |    0.005 |   0.000345 |   0.000748\n",
      "    23 |      0.056 |    133.704 |      0.252 |      0.524 |     128.00 |     128.00 |    6.656 |    0.000 |    0.005 |   0.000783 |   0.000419\n",
      "    24 |      0.020 |    134.092 |      0.252 |      0.524 |     128.00 |     256.00 |    4.786 |    0.000 |    0.005 |   0.000274 |   0.000159\n",
      "    25 |      0.029 |    144.511 |      0.263 |      0.526 |     256.00 |     256.00 |    4.331 |    0.000 |    0.004 |   0.000096 |   0.000202\n",
      "    26 |      0.016 |    144.515 |      0.263 |      0.527 |     256.00 |     256.00 |    5.635 |    0.000 |    0.004 |   0.000221 |   0.000113\n",
      "    27 |      0.006 |    144.595 |      0.263 |      0.527 |     256.00 |     512.00 |    4.082 |    0.000 |    0.004 |   0.000080 |   0.000043\n",
      "    28 |      0.008 |    150.618 |      0.269 |      0.529 |     512.00 |          / |    3.968 |    0.000 |    0.004 |   0.000026 |   0.000057\n",
      "early stop!\n",
      "Terminated (optimal) in 29 iterations.\n",
      "One optimization by ADMM finished. Elapsed time: 4.26 minutes.\n",
      "\n",
      "\n",
      "stage 2 finished. Elapsed time: 119.40 minutes.\n",
      "\n",
      "GLRM fitting finished. Elapsed time: 143.32 minutes.\n",
      "\n",
      "\n",
      "Post-processing estimated cell-type proportion theta...\n",
      "hard thresholding small theta values with threshold 0\n",
      "\n",
      "\n",
      "cell type deconvolution finished. Estimate results saved in /home/exouser/Spatial/celltype_proportions.csv. Elapsed time: 2.60 hours.\n",
      "\n",
      "\n",
      "######### No imputation #########\n",
      "\n",
      "\n",
      "whole pipeline finished. Total elapsed time: 2.60 hours.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "CompletedProcess(args='runDeconvolution -q sim_seq_based_spatial_spot_nUMI.csv                           -r scRNA_data_full.csv                           -c ref_scRNA_cell_celltype.csv                           -a sim_spatial_spot_adjacency_matrix.csv                           --n_marker_per_cmp 20                           -n 64                           --use_cvae false ', returncode=0)"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import subprocess\n",
    "\n",
    "cmd = '''runDeconvolution -q sim_seq_based_spatial_spot_nUMI.csv \\\n",
    "                          -r scRNA_data_full.csv \\\n",
    "                          -c ref_scRNA_cell_celltype.csv \\\n",
    "                          -a sim_spatial_spot_adjacency_matrix.csv \\\n",
    "                          --n_marker_per_cmp 20 \\\n",
    "                          -n 64 \\\n",
    "                          --use_cvae false \\\n",
    "'''\n",
    "\n",
    "subprocess.run(cmd, check=True, text=True, shell=True)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
