{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "db496b92",
   "metadata": {},
   "source": [
    "# Run *SDePER* on sequencing-based simulated data: Scenario 1 + Spatial data as reference + WITHOUT sparsity penalty + WITHOUT spatial correlation constraint\n",
    "\n",
    "In this Notebook we run ablation test on SDePER on simulated data. For generating **sequencing-based** simulated data via coarse-graining procedure please refer [generate_simulated_spatial_data.nb.html](https://rawcdn.githack.com/az7jh2/SDePER_Analysis/c963d08f74f4591c2ef6f132177795297793d878/Simulation_seq_based/Generate_simulation_data/generate_simulated_spatial_data.nb.html) in [Generate_simulation_data](https://github.com/az7jh2/SDePER_Analysis/tree/main/Simulation_seq_based/Generate_simulation_data) folder.\n",
    "\n",
    "**Scenario 1** means the reference data for deconvolution includes all single cells with the **matched 12 cell types**.\n",
    "\n",
    "**Spatial data as reference** means the reference data is actually the [GSE102827](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE102827) scRNA-seq data which is used to generate the simulated data, therefore it's **free of platform effect**.\n",
    "\n",
    "**WITHOUT sparsity penalty** means we fit the graph Laplacian regularized model without the sparsity penalty, specifically omitting Adaptive LASSO regularization (by setting the command option `--lambda_r` to 0).\n",
    "\n",
    "**WITHOUT spatial correlation constraint** means we fit the graph Laplacian regularized model the spatial correlation constraint, essentially removing Laplacian regularization (by setting the command option `--lambda_g` to 0).\n",
    "\n",
    "==================================================================================================================\n",
    "\n",
    "So here we use the **4 input files** as shown below:\n",
    "\n",
    "!!! NOTE here we directly use CVAE transformed spatial transcriptomic data and scRNA-seq marker gene profile as input, which can be found in [diagnosis file](https://github.com/az7jh2/SDePER_Analysis/blob/main/Simulation_seq_based/Run_SDePER_on_simulation_data/Scenario_1/ref_spatial/S1_ref_spatial_SDePER_WITH_CVAE_diagnosis.tar) generated in baseline run [S1_ref_spatial_SDePER_WITH_CVAE.ipynb](https://github.com/az7jh2/SDePER_Analysis/blob/main/Simulation_seq_based/Run_SDePER_on_simulation_data/Scenario_1/ref_spatial/S1_ref_spatial_SDePER_WITH_CVAE.ipynb) !!!\n",
    "\n",
    "1. nUMI-like counts of transformed spatial transcriptomic data (spots × genes): `spatial_spots_transformToscRNA_decoded.csv`\n",
    "2. reference cell type specific marker gene expression (cell types × genes): modified `scRNA_decoded_avg_exp_bycelltypes.csv` to include only selected marker genes in `redo_DE_celltype_markers.csv`\n",
    "4. adjacency matrix of spots in simulated spatial transcriptomic data (spots × spots): [sim_spatial_spot_adjacency_matrix.csv](https://github.com/az7jh2/SDePER_Analysis/blob/main/Simulation/Generate_simulation_data/sim_spatial_spot_adjacency_matrix.csv)\n",
    "\n",
    "==================================================================================================================\n",
    "\n",
    "SDePER settings are the same as baseline run [S1_ref_spatial_SDePER_WITH_CVAE.ipynb](https://github.com/az7jh2/SDePER_Analysis/blob/main/Simulation_seq_based/Run_SDePER_on_simulation_data/Scenario_1/ref_spatial/S1_ref_spatial_SDePER_WITH_CVAE.ipynb), and we discarded unneeded command-line options further disabled the additive platform effect term by manually adjusted the source code:\n",
    "\n",
    "* number of used CPU cores `n_core`: 64\n",
    "\n",
    "ALL other options are left as default.\n",
    "\n",
    "**For ablation test, set hyper-parameter for Adaptive Lasso `lambda_r` as 0, and hyper-parameter for Graph Laplacian Constrain `lambda_g` as 0.**\n",
    "\n",
    "==================================================================================================================\n",
    "\n",
    "the `bash` command to start cell type deconvolution is\n",
    "\n",
    "`runDeconvolution -q spatial_spots_transformToscRNA_decoded.csv -m scRNA_decoded_avg_exp_bycelltypes.csv -a sim_spatial_spot_adjacency_matrix.csv -n 64 --lambda_r 0 --lambda_g 0`\n",
    "\n",
    "Note this Notebook uses **SDePER v1.2.1**. Cell type deconvolution result is renamed as [S1_ref_spatial_SDePER_Ablation_NO_LASSO_Laplacian_celltype_proportions.csv](https://github.com/az7jh2/SDePER_Analysis/blob/main/Ablation/Ablation_simulation_seq_based/S1_ref_spatial_SDePER_Ablation_NO_LASSO_Laplacian_celltype_proportions.csv)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "1a38dcbd-906e-46aa-8aee-7a1dd9c66d6f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "got 573 genes for 12 cell types\n",
      "got 382 marker genes for downstream analysis\n"
     ]
    }
   ],
   "source": [
    "# modify cell type specific marker gene expression profile to included only selected markers\n",
    "import pandas as pd\n",
    "\n",
    "marker_df = pd.read_csv('scRNA_decoded_avg_exp_bycelltypes.csv', index_col=0)\n",
    "print(f'got {marker_df.shape[1]} genes for {marker_df.shape[0]} cell types')\n",
    "\n",
    "de_gene_df = pd.read_csv('redo_DE_celltype_markers.csv')\n",
    "de_gene_list = list(de_gene_df.loc[de_gene_df['selected']==1, 'gene'].unique())\n",
    "print(f'got {len(de_gene_list)} marker genes for downstream analysis')\n",
    "\n",
    "marker_df[de_gene_list].to_csv('scRNA_decoded_avg_exp_bycelltypes.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "df9e3939",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "SDePER (Spatial Deconvolution method with Platform Effect Removal) v1.2.1\n",
      "\n",
      "WARNING: building CVAE requires reference scRNA-seq data and corresponding cell-type annotation specified! But at least one of them is not specified. use_cvae will be reset to `False`!\n",
      "\n",
      "running options:\n",
      "spatial_file: /home/exouser/Spatial/spatial_spots_transformToscRNA_decoded.csv\n",
      "ref_file: None\n",
      "ref_celltype_file: None\n",
      "marker_file: /home/exouser/Spatial/scRNA_decoded_avg_exp_bycelltypes.csv\n",
      "loc_file: None\n",
      "A_file: /home/exouser/Spatial/sim_spatial_spot_adjacency_matrix.csv\n",
      "n_cores: 64\n",
      "threshold: 0\n",
      "use_cvae: False\n",
      "use_imputation: False\n",
      "diagnosis: False\n",
      "verbose: True\n",
      "use_fdr: True\n",
      "p_val_cutoff: 0.05\n",
      "fc_cutoff: 1.2\n",
      "pct1_cutoff: 0.3\n",
      "pct2_cutoff: 0.1\n",
      "sortby_fc: True\n",
      "n_marker_per_cmp: 20\n",
      "filter_cell: True\n",
      "filter_gene: True\n",
      "n_hv_gene: 200\n",
      "n_pseudo_spot: 500000\n",
      "pseudo_spot_min_cell: 2\n",
      "pseudo_spot_max_cell: 8\n",
      "seq_depth_scaler: 10000\n",
      "cvae_input_scaler: 10\n",
      "cvae_init_lr: 0.01\n",
      "num_hidden_layer: 2\n",
      "use_batch_norm: True\n",
      "cvae_train_epoch: 500\n",
      "use_spatial_pseudo: False\n",
      "redo_de: True\n",
      "seed: 383\n",
      "lambda_r: 0.0\n",
      "lambda_g: 0.0\n",
      "diameter: 200\n",
      "impute_diameter: [160, 114, 80]\n",
      "\n",
      "\n",
      "######### Preprocessing... #########\n",
      "\n",
      "building CVAE skipped...\n",
      "\n",
      "read spatial data from file /home/exouser/Spatial/spatial_spots_transformToscRNA_decoded.csv\n",
      "total 581 spots; 573 genes\n",
      "\n",
      "filtering genes present in <3 spots: 2 genes removed\n",
      "\n",
      "user provided marker gene profile, DE will be skipped...\n",
      "\n",
      "read 382 marker genes from user specified marker gene file\n",
      "from user specified marker gene expression use 382 marker genes overlapped with spatial + scRNA-seq data\n",
      "\n",
      "gene filtering before modeling...\n",
      "all genes passed filtering\n",
      "\n",
      "spot filtering before modeling...\n",
      "all spots passed filtering\n",
      "\n",
      "\n",
      "######### Start GLRM modeling... #########\n",
      "\n",
      "GLRM settings:\n",
      "use SciPy minimize method:  L-BFGS-B\n",
      "global optimization turned off, local minimum will be used in GLRM\n",
      "use hybrid version of GLRM\n",
      "Numba detected total 64 available CPU cores. Use 64 CPU cores\n",
      "use 2001 points to calculate the heavy-tail density\n",
      "use weight threshold for Adaptive Lasso:  0.001\n",
      "total 311 unique nUMIs, min: 0, max: 1248\n",
      "\n",
      "Build graph: \n",
      " Graph with 581 nodes and 1029 edges\n",
      "\n",
      "estimation of gene-specific platform effect gamma_g is skipped as already using CVAE to adjust platform effect\n",
      "\n",
      "\n",
      "Start GLRM fitting...\n",
      "\n",
      "first estimate MLE theta and corresponding e^alpha and sigma^2...\n",
      "\n",
      "GLRM model initialization...\n",
      "calculate MLE theta and sigma^2 iteratively...\n",
      "  iter | time_opt | time_sig | sigma2\n",
      "     0 |   17.649 |    3.566 |  0.377\n",
      "     1 |    9.876 |    2.583 |  0.224\n",
      "     2 |    8.718 |    2.775 |  0.183\n",
      "     3 |    7.569 |    2.224 |  0.172\n",
      "     4 |    6.509 |    1.915 |  0.169\n",
      "     5 |    5.154 |    1.585 |  0.168\n",
      "MLE theta and sigma^2 calculation finished. Elapsed time: 1.17 minutes.\n",
      "MLE theta estimation finished. Elapsed time: 1.17 minutes.\n",
      "\n",
      "calculate weights of Adaptive Lasso...\n",
      "\n",
      "Stage 1: variable selection using Adaptive Lasso starts with the MLE theta and e^alpha, using already estimated sigma^2 and gamma_g...\n",
      "specified hyper-parameter for Adaptive Lasso is: 0.0\n",
      "\n",
      "start ADMM iteration...\n",
      "  iter |  res_pri_n | res_dual_n |    eps_pri |   eps_dual |        rho |    new_rho | time_opt | time_reg | time_lap | tilde_RMSE |   hat_RMSE\n",
      "     0 |      0.004 |      0.004 |      0.118 |      0.118 |       1.00 |          / |    3.530 |    0.000 |    0.004 |   0.000034 |   0.000017\n",
      "Terminated (optimal) in 1 iterations.\n",
      "One optimization by ADMM finished. Elapsed time: 0.06 minutes.\n",
      "\n",
      "Stage 1 variable selection finished. Elapsed time: 0.06 minutes.\n",
      "\n",
      "Stage 2: final theta estimation with Graph Laplacian Constrain using already estimated sigma^2 and gamma_g\n",
      "specified hyper-parameter for Graph Laplacian Constrain is: 0.0\n",
      "\n",
      "start ADMM iteration...\n",
      "  iter |  res_pri_n | res_dual_n |    eps_pri |   eps_dual |        rho |    new_rho | time_opt | time_reg | time_lap | tilde_RMSE |   hat_RMSE\n",
      "     0 |      8.504 |      8.504 |      0.127 |      0.127 |       1.00 |       1.00 |    8.519 |    0.000 |    0.003 |   0.133793 |   0.066897\n",
      "     1 |      0.000 |      8.504 |      0.127 |      0.127 |       1.00 |          / |    1.448 |    0.000 |    0.003 |   0.000001 |   0.000000\n",
      "early stop!\n",
      "Terminated (optimal) in 2 iterations.\n",
      "One optimization by ADMM finished. Elapsed time: 0.17 minutes.\n",
      "\n",
      "\n",
      "stage 2 finished. Elapsed time: 0.17 minutes.\n",
      "\n",
      "GLRM fitting finished. Elapsed time: 1.39 minutes.\n",
      "\n",
      "\n",
      "Post-processing estimated cell-type proportion theta...\n",
      "hard thresholding small theta values with threshold 0\n",
      "\n",
      "\n",
      "cell type deconvolution finished. Estimate results saved in /home/exouser/Spatial/celltype_proportions.csv. Elapsed time: 0.02 hours.\n",
      "\n",
      "\n",
      "######### No imputation #########\n",
      "\n",
      "\n",
      "whole pipeline finished. Total elapsed time: 0.02 hours.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "CompletedProcess(args='runDeconvolution -q spatial_spots_transformToscRNA_decoded.csv                           -m scRNA_decoded_avg_exp_bycelltypes.csv                           -a sim_spatial_spot_adjacency_matrix.csv                           -n 64                           --lambda_r 0                           --lambda_g 0\\n', returncode=0)"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import subprocess\n",
    "\n",
    "cmd = '''runDeconvolution -q spatial_spots_transformToscRNA_decoded.csv \\\n",
    "                          -m scRNA_decoded_avg_exp_bycelltypes.csv \\\n",
    "                          -a sim_spatial_spot_adjacency_matrix.csv \\\n",
    "                          -n 64 \\\n",
    "                          --lambda_r 0 \\\n",
    "                          --lambda_g 0\n",
    "'''\n",
    "\n",
    "subprocess.run(cmd, check=True, text=True, shell=True)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
