{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "02f03238",
   "metadata": {},
   "source": [
    "# Run *SDePER* on Real data: Melanoma\n",
    "\n",
    "In this Notebook we run SDePER on one Real Dataset -- human **Melanoma**. \n",
    "\n",
    "For downloading and preprocessing original spatial and reference scRNA-seq data for cell type deconvolution please refer [Melanoma_preprocess.nb.html](https://rawcdn.githack.com/az7jh2/SDePER_Analysis/0a429908645a665c1f9d345df013d5b9fcde20b3/RealData/Melanoma/Melanoma_preprocess.nb.html).\n",
    "\n",
    "==================================================================================================================\n",
    "\n",
    "So here we use **5 input files** as shown below:\n",
    "\n",
    "1. raw nUMI counts of spatial transcriptomic data (spots × genes): [Melanoma_spatial_spot_nUMI.csv](https://github.com/az7jh2/SDePER_Analysis/blob/main/RealData/Melanoma/Melanoma_spatial_spot_nUMI.csv)\n",
    "2. raw nUMI counts of reference scRNA-seq data (cells × genes): `Melanoma_ref_scRNA_cell_nUMI.csv`, just decompress the gzipped file [Melanoma_ref_scRNA_cell_nUMI.csv.gz](https://github.com/az7jh2/SDePER_Analysis/blob/main/RealData/Melanoma/Melanoma_ref_scRNA_cell_nUMI.csv.gz)\n",
    "3. cell type annotations for cells of selected 7 cell types in reference scRNA-seq data (cells × 1): [Melanoma_ref_scRNA_cell_celltype.csv](https://github.com/az7jh2/SDePER_Analysis/blob/main/RealData/Melanoma/Melanoma_ref_scRNA_cell_celltype.csv)\n",
    "4. adjacency matrix of spots in spatial transcriptomic data (spots × spots): [Melanoma_spatial_spot_adjacency_matrix.csv](https://github.com/az7jh2/SDePER_Analysis/blob/main/RealData/Melanoma/Melanoma_spatial_spot_adjacency_matrix.csv)\n",
    "5. row/column integer index (x,y) of spatial spots (spots * 2): [Melanoma_spatial_spot_loc.csv](https://github.com/az7jh2/SDePER_Analysis/blob/main/RealData/Melanoma/Melanoma_spatial_spot_loc.csv)\n",
    "\n",
    "==================================================================================================================\n",
    "\n",
    "SDePER settings are:\n",
    "\n",
    "* number of included highly variable genes `n_hv_gene`: 300\n",
    "* number of selected TOP marker genes for each comparison in Differential `n_marker_per_cmp`: 10\n",
    "* minimum value of cells in one pseudo-spot for building CVAE `pseudo_spot_min_cell`: 5\n",
    "* maximum value of cells in one pseudo-spot for building CVAE `pseudo_spot_max_cell`: 40\n",
    "* seed for random values `seed`: 3\n",
    "* number of used CPU cores `n_core`: 32\n",
    "* whether to perform imputation `use_imputation`: true\n",
    "\n",
    "ALL other options are left as default.\n",
    "\n",
    "==================================================================================================================\n",
    "\n",
    "the `bash` command to start cell type deconvolution is\n",
    "\n",
    "`runDeconvolution -q Melanoma_spatial_spot_nUMI.csv -r Melanoma_ref_scRNA_cell_nUMI.csv -c Melanoma_ref_scRNA_cell_celltype.csv -a Melanoma_spatial_spot_adjacency_matrix.csv -l Melanoma_spatial_spot_loc.csv --n_hv_gene 300 --n_marker_per_cmp 10 --pseudo_spot_min_cell 5 --pseudo_spot_max_cell 40 --seed 3 -n 32 --use_imputation true`\n",
    "\n",
    "Note this Notebook uses **SDePER v1.0.1**. Cell type deconvolution result is renamed as [Melanoma_SDePER_celltype_proportions.csv](https://github.com/az7jh2/SDePER_Analysis/tree/main/RealData/Melanoma/Melanoma_SDePER_celltype_proportions.csv). Imputation results are compressed into into one zipped file and deposited to folder [Melanoma](https://zenodo.org/record/8334656/files/Melanoma.zip) in [10.5281/zenodo.8334655](https://doi.org/10.5281/zenodo.8334655)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "f43766d1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "SDePER (Spatial Deconvolution method with Platform Effect Removal) v1.0.1\n",
      "\n",
      "\n",
      "running options:\n",
      "spatial_file: /home/exouser/Spatial/Melanoma_spatial_spot_nUMI.csv\n",
      "ref_file: /home/exouser/Spatial/Melanoma_ref_scRNA_cell_nUMI.csv\n",
      "ref_celltype_file: /home/exouser/Spatial/Melanoma_ref_scRNA_cell_celltype.csv\n",
      "marker_file: None\n",
      "loc_file: /home/exouser/Spatial/Melanoma_spatial_spot_loc.csv\n",
      "A_file: /home/exouser/Spatial/Melanoma_spatial_spot_adjacency_matrix.csv\n",
      "n_cores: 32\n",
      "lambda_r: [0.1, 0.268, 0.72, 1.931, 5.179, 13.895, 37.276, 100.0]\n",
      "lambda_g: [0.1, 0.268, 0.72, 1.931, 5.179, 13.895, 37.276, 100.0]\n",
      "use_cvae: True\n",
      "threshold: 0\n",
      "n_hv_gene: 300\n",
      "n_marker_per_cmp: 10\n",
      "pseudo_spot_min_cell: 5\n",
      "pseudo_spot_max_cell: 40\n",
      "seq_depth_scaler: 10000\n",
      "cvae_input_scaler: 10\n",
      "cvae_init_lr: 0.003\n",
      "redo_de: True\n",
      "seed: 3\n",
      "diagnosis: False\n",
      "verbose: True\n",
      "use_imputation: True\n",
      "diameter: 200\n",
      "impute_diameter: [160, 114, 80]\n",
      "\n",
      "\n",
      "######### Preprocessing... #########\n",
      "\n",
      "######### First build CVAE... #########\n",
      "\n",
      "read spatial data from file /home/exouser/Spatial/Melanoma_spatial_spot_nUMI.csv\n",
      "total 293 spots; 16148 genes\n",
      "\n",
      "read scRNA-seq data from file /home/exouser/Spatial/Melanoma_ref_scRNA_cell_nUMI.csv\n",
      "total 2495 cells; 23686 genes\n",
      "read scRNA-seq cell-type annotation from file /home/exouser/Spatial/Melanoma_ref_scRNA_cell_celltype.csv\n",
      "total 7 cell-types\n",
      "subset cells with cell-type annotation, finally keep 2495 cells; 23686 genes\n",
      "\n",
      "total 14522 overlapped genes\n",
      "\n",
      "identify 300 highly variable genes from scRNA-seq data...\n",
      "\n",
      "identify cell-type marker genes...\n",
      "no marker gene profile provided. Perform DE to get cell-type marker genes on scRNA-seq data...\n",
      "\n",
      "Differential analysis across cell-types on scRNA-seq data...\n",
      "finally selected 280 cell-type marker genes\n",
      "\n",
      "\n",
      "use union of highly variable gene list and cell-type marker gene list derived from scRNA-seq data, finally get 534 genes for downstream analysis\n",
      "\n",
      "start CVAE building...\n",
      "\n",
      "generate pseudo-spots containing 5 to 40 cells from scRNA-seq cells...\n",
      "generate 164080 pseudo-spots for training and 41020 pseudo-spots for validation\n",
      "scaling inputs to range 0 to 10\n",
      "in training -- spatial spots : pseudo-spots = 293 : 166575\n",
      "\n",
      "Start training...\n",
      "\n",
      "Train on 166868 samples, validate on 41020 samples\n",
      "Epoch 1/1000\n",
      "166868/166868 [==============================] - 3s 15us/sample - loss: 196.0020 - val_loss: 51.1262 - lr: 0.0030\n",
      "Epoch 2/1000\n",
      "166868/166868 [==============================] - 2s 10us/sample - loss: 160.2350 - val_loss: 48.2872 - lr: 0.0030\n",
      "Epoch 3/1000\n",
      "166868/166868 [==============================] - 2s 9us/sample - loss: 151.1435 - val_loss: 45.9141 - lr: 0.0030\n",
      "Epoch 4/1000\n",
      "166868/166868 [==============================] - 1s 9us/sample - loss: 144.0527 - val_loss: 43.6343 - lr: 0.0030\n",
      "Epoch 5/1000\n",
      "166868/166868 [==============================] - 1s 8us/sample - loss: 138.1909 - val_loss: 41.1883 - lr: 0.0030\n",
      "Epoch 6/1000\n",
      "166868/166868 [==============================] - 1s 9us/sample - loss: 133.0362 - val_loss: 38.9653 - lr: 0.0030\n",
      "Epoch 7/1000\n",
      "166868/166868 [==============================] - 1s 9us/sample - loss: 128.7477 - val_loss: 37.2812 - lr: 0.0030\n",
      "Epoch 8/1000\n",
      "166868/166868 [==============================] - 1s 9us/sample - loss: 125.3912 - val_loss: 35.8204 - lr: 0.0030\n",
      "Epoch 9/1000\n",
      "166868/166868 [==============================] - 1s 9us/sample - loss: 122.6177 - val_loss: 34.6131 - lr: 0.0030\n",
      "Epoch 10/1000\n",
      "166868/166868 [==============================] - 1s 9us/sample - loss: 120.0179 - val_loss: 33.8590 - lr: 0.0030\n",
      "Epoch 11/1000\n",
      "166868/166868 [==============================] - 1s 9us/sample - loss: 118.2675 - val_loss: 33.0173 - lr: 0.0030\n",
      "Epoch 12/1000\n",
      "166868/166868 [==============================] - 1s 9us/sample - loss: 116.4938 - val_loss: 32.2120 - lr: 0.0030\n",
      "Epoch 13/1000\n",
      "166868/166868 [==============================] - 1s 9us/sample - loss: 114.7203 - val_loss: 31.7333 - lr: 0.0030\n",
      "Epoch 14/1000\n",
      "166868/166868 [==============================] - 1s 9us/sample - loss: 113.4490 - val_loss: 30.9636 - lr: 0.0030\n",
      "Epoch 15/1000\n",
      "166868/166868 [==============================] - 1s 9us/sample - loss: 111.8172 - val_loss: 30.3037 - lr: 0.0030\n",
      "Epoch 16/1000\n",
      "166868/166868 [==============================] - 1s 9us/sample - loss: 110.5249 - val_loss: 29.7141 - lr: 0.0030\n",
      "Epoch 17/1000\n",
      "166868/166868 [==============================] - 1s 9us/sample - loss: 109.3142 - val_loss: 29.3918 - lr: 0.0030\n",
      "Epoch 18/1000\n",
      "166868/166868 [==============================] - 1s 9us/sample - loss: 108.3114 - val_loss: 28.7863 - lr: 0.0030\n",
      "Epoch 19/1000\n",
      "166868/166868 [==============================] - 1s 9us/sample - loss: 107.2275 - val_loss: 28.4682 - lr: 0.0030\n",
      "Epoch 20/1000\n",
      "166868/166868 [==============================] - 1s 9us/sample - loss: 106.4363 - val_loss: 27.8479 - lr: 0.0030\n",
      "Epoch 21/1000\n",
      "166868/166868 [==============================] - 1s 9us/sample - loss: 105.3813 - val_loss: 27.8062 - lr: 0.0030\n",
      "Epoch 22/1000\n",
      "166868/166868 [==============================] - 1s 9us/sample - loss: 105.0220 - val_loss: 27.4483 - lr: 0.0030\n",
      "Epoch 23/1000\n",
      "166868/166868 [==============================] - 1s 9us/sample - loss: 104.1669 - val_loss: 26.9510 - lr: 0.0030\n",
      "Epoch 24/1000\n",
      "166868/166868 [==============================] - 1s 8us/sample - loss: 103.2928 - val_loss: 26.6908 - lr: 0.0030\n",
      "Epoch 25/1000\n",
      "166868/166868 [==============================] - 2s 9us/sample - loss: 102.7998 - val_loss: 26.3614 - lr: 0.0030\n",
      "Epoch 26/1000\n",
      "166868/166868 [==============================] - 1s 9us/sample - loss: 102.0722 - val_loss: 26.0787 - lr: 0.0030\n",
      "Epoch 27/1000\n",
      "166868/166868 [==============================] - 1s 9us/sample - loss: 101.4375 - val_loss: 25.8689 - lr: 0.0030\n",
      "Epoch 28/1000\n",
      "166868/166868 [==============================] - 1s 9us/sample - loss: 101.0575 - val_loss: 25.5734 - lr: 0.0030\n",
      "Epoch 29/1000\n",
      "166868/166868 [==============================] - 1s 9us/sample - loss: 100.5253 - val_loss: 25.3863 - lr: 0.0030\n",
      "Epoch 30/1000\n",
      "166868/166868 [==============================] - 1s 9us/sample - loss: 100.0883 - val_loss: 25.1100 - lr: 0.0030\n",
      "Epoch 31/1000\n",
      "166868/166868 [==============================] - 2s 9us/sample - loss: 99.6410 - val_loss: 25.0110 - lr: 0.0030\n",
      "Epoch 32/1000\n",
      "166868/166868 [==============================] - 1s 9us/sample - loss: 99.3141 - val_loss: 24.7577 - lr: 0.0030\n",
      "Epoch 33/1000\n",
      "166868/166868 [==============================] - 1s 9us/sample - loss: 98.9325 - val_loss: 24.5364 - lr: 0.0030\n",
      "Epoch 34/1000\n",
      "166868/166868 [==============================] - 2s 9us/sample - loss: 98.4405 - val_loss: 24.3626 - lr: 0.0030\n",
      "Epoch 35/1000\n",
      "166868/166868 [==============================] - 2s 9us/sample - loss: 98.0962 - val_loss: 24.2723 - lr: 0.0030\n",
      "Epoch 36/1000\n",
      "166868/166868 [==============================] - 2s 9us/sample - loss: 97.7430 - val_loss: 24.0666 - lr: 0.0030\n",
      "Epoch 37/1000\n",
      "166868/166868 [==============================] - 2s 9us/sample - loss: 97.4480 - val_loss: 23.9319 - lr: 0.0030\n",
      "Epoch 38/1000\n",
      "166868/166868 [==============================] - 2s 9us/sample - loss: 97.0772 - val_loss: 23.7896 - lr: 0.0030\n",
      "Epoch 39/1000\n",
      "166868/166868 [==============================] - 2s 9us/sample - loss: 96.7913 - val_loss: 23.6679 - lr: 0.0030\n",
      "Epoch 40/1000\n",
      "166868/166868 [==============================] - 2s 9us/sample - loss: 96.3457 - val_loss: 23.6332 - lr: 0.0030\n",
      "Epoch 41/1000\n",
      "166868/166868 [==============================] - 2s 9us/sample - loss: 96.1345 - val_loss: 23.5192 - lr: 0.0030\n",
      "Epoch 42/1000\n",
      "166868/166868 [==============================] - 2s 9us/sample - loss: 95.7302 - val_loss: 23.4064 - lr: 0.0030\n",
      "Epoch 43/1000\n",
      "166868/166868 [==============================] - 2s 9us/sample - loss: 95.4782 - val_loss: 23.3318 - lr: 0.0030\n",
      "Epoch 44/1000\n",
      "166868/166868 [==============================] - 2s 9us/sample - loss: 95.1705 - val_loss: 23.2274 - lr: 0.0030\n",
      "Epoch 45/1000\n",
      "166868/166868 [==============================] - 2s 9us/sample - loss: 94.9061 - val_loss: 23.1675 - lr: 0.0030\n",
      "Epoch 46/1000\n",
      "166868/166868 [==============================] - 2s 9us/sample - loss: 94.6658 - val_loss: 23.0685 - lr: 0.0030\n",
      "Epoch 47/1000\n",
      "166868/166868 [==============================] - 2s 9us/sample - loss: 94.2613 - val_loss: 23.0035 - lr: 0.0030\n",
      "Epoch 48/1000\n",
      "166868/166868 [==============================] - 2s 9us/sample - loss: 94.0076 - val_loss: 22.9233 - lr: 0.0030\n",
      "Epoch 49/1000\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "166868/166868 [==============================] - 2s 9us/sample - loss: 93.6491 - val_loss: 22.8037 - lr: 0.0030\n",
      "Epoch 50/1000\n",
      "166868/166868 [==============================] - 2s 9us/sample - loss: 93.2764 - val_loss: 22.8045 - lr: 0.0030\n",
      "Epoch 51/1000\n",
      "166868/166868 [==============================] - 2s 9us/sample - loss: 92.9578 - val_loss: 22.6965 - lr: 0.0030\n",
      "Epoch 52/1000\n",
      "166868/166868 [==============================] - 2s 9us/sample - loss: 92.6443 - val_loss: 22.6768 - lr: 0.0030\n",
      "Epoch 53/1000\n",
      "166868/166868 [==============================] - 2s 9us/sample - loss: 92.3823 - val_loss: 22.5972 - lr: 0.0030\n",
      "Epoch 54/1000\n",
      "166868/166868 [==============================] - 2s 9us/sample - loss: 92.0571 - val_loss: 22.5105 - lr: 0.0030\n",
      "Epoch 55/1000\n",
      "166868/166868 [==============================] - 2s 9us/sample - loss: 91.5985 - val_loss: 22.4705 - lr: 0.0030\n",
      "Epoch 56/1000\n",
      "166868/166868 [==============================] - 2s 9us/sample - loss: 91.4056 - val_loss: 22.3807 - lr: 0.0030\n",
      "Epoch 57/1000\n",
      "166868/166868 [==============================] - 2s 9us/sample - loss: 91.1129 - val_loss: 22.3498 - lr: 0.0030\n",
      "Epoch 58/1000\n",
      "166868/166868 [==============================] - 2s 9us/sample - loss: 90.9028 - val_loss: 22.2655 - lr: 0.0030\n",
      "Epoch 59/1000\n",
      "166868/166868 [==============================] - 2s 10us/sample - loss: 90.4822 - val_loss: 22.2666 - lr: 0.0030\n",
      "Epoch 60/1000\n",
      "166868/166868 [==============================] - 2s 9us/sample - loss: 90.2721 - val_loss: 22.2078 - lr: 0.0030\n",
      "Epoch 61/1000\n",
      "166868/166868 [==============================] - 2s 9us/sample - loss: 89.8921 - val_loss: 22.1634 - lr: 0.0030\n",
      "Epoch 62/1000\n",
      "166868/166868 [==============================] - 2s 9us/sample - loss: 89.5970 - val_loss: 22.1389 - lr: 0.0030\n",
      "Epoch 63/1000\n",
      "166868/166868 [==============================] - 2s 9us/sample - loss: 89.2951 - val_loss: 22.1049 - lr: 0.0030\n",
      "Epoch 64/1000\n",
      "166868/166868 [==============================] - 2s 10us/sample - loss: 88.9551 - val_loss: 22.0890 - lr: 0.0030\n",
      "Epoch 65/1000\n",
      "166868/166868 [==============================] - 2s 9us/sample - loss: 88.6895 - val_loss: 22.0538 - lr: 0.0030\n",
      "Epoch 66/1000\n",
      "166868/166868 [==============================] - 2s 9us/sample - loss: 88.3071 - val_loss: 22.0605 - lr: 0.0030\n",
      "Epoch 67/1000\n",
      "166868/166868 [==============================] - 2s 9us/sample - loss: 88.1158 - val_loss: 22.0275 - lr: 0.0030\n",
      "Epoch 68/1000\n",
      "166868/166868 [==============================] - 2s 9us/sample - loss: 87.7854 - val_loss: 22.0377 - lr: 0.0030\n",
      "Epoch 69/1000\n",
      "166868/166868 [==============================] - 2s 10us/sample - loss: 87.5220 - val_loss: 21.9944 - lr: 0.0030\n",
      "Epoch 70/1000\n",
      "166868/166868 [==============================] - 2s 10us/sample - loss: 87.1789 - val_loss: 22.0343 - lr: 0.0030\n",
      "Epoch 71/1000\n",
      "166868/166868 [==============================] - 2s 10us/sample - loss: 86.8697 - val_loss: 21.9951 - lr: 0.0030\n",
      "Epoch 72/1000\n",
      "166868/166868 [==============================] - 2s 9us/sample - loss: 86.6079 - val_loss: 21.9839 - lr: 0.0030\n",
      "Epoch 73/1000\n",
      "166868/166868 [==============================] - 2s 10us/sample - loss: 86.3520 - val_loss: 22.0063 - lr: 0.0030\n",
      "Epoch 74/1000\n",
      "166868/166868 [==============================] - 2s 10us/sample - loss: 86.1884 - val_loss: 21.9921 - lr: 0.0030\n",
      "Epoch 75/1000\n",
      "166868/166868 [==============================] - 2s 10us/sample - loss: 85.7198 - val_loss: 21.9944 - lr: 0.0030\n",
      "Epoch 76/1000\n",
      "166868/166868 [==============================] - 2s 10us/sample - loss: 85.5154 - val_loss: 21.9799 - lr: 0.0030\n",
      "Epoch 77/1000\n",
      "166868/166868 [==============================] - 2s 10us/sample - loss: 85.1577 - val_loss: 22.0144 - lr: 0.0030\n",
      "Epoch 78/1000\n",
      "166868/166868 [==============================] - 2s 10us/sample - loss: 84.8172 - val_loss: 21.9897 - lr: 0.0030\n",
      "Epoch 79/1000\n",
      "166868/166868 [==============================] - 2s 10us/sample - loss: 84.6881 - val_loss: 22.0164 - lr: 0.0030\n",
      "Epoch 80/1000\n",
      "166868/166868 [==============================] - 2s 10us/sample - loss: 84.3714 - val_loss: 22.0157 - lr: 0.0030\n",
      "Epoch 81/1000\n",
      "166868/166868 [==============================] - 2s 10us/sample - loss: 83.9684 - val_loss: 22.0441 - lr: 0.0030\n",
      "Epoch 82/1000\n",
      "166868/166868 [==============================] - 2s 10us/sample - loss: 83.6550 - val_loss: 22.0506 - lr: 0.0030\n",
      "Epoch 83/1000\n",
      "166868/166868 [==============================] - 2s 9us/sample - loss: 83.4807 - val_loss: 22.0875 - lr: 0.0030\n",
      "Epoch 84/1000\n",
      "166868/166868 [==============================] - 2s 10us/sample - loss: 83.1750 - val_loss: 22.0816 - lr: 0.0030\n",
      "Epoch 85/1000\n",
      "166868/166868 [==============================] - 2s 9us/sample - loss: 82.9816 - val_loss: 22.0487 - lr: 0.0030\n",
      "Epoch 86/1000\n",
      "166868/166868 [==============================] - 2s 10us/sample - loss: 82.6357 - val_loss: 22.0947 - lr: 0.0030\n",
      "Epoch 87/1000\n",
      "166868/166868 [==============================] - 2s 9us/sample - loss: 82.3517 - val_loss: 22.0950 - lr: 0.0030\n",
      "Epoch 88/1000\n",
      "166868/166868 [==============================] - 2s 9us/sample - loss: 82.1093 - val_loss: 22.0952 - lr: 0.0030\n",
      "Epoch 89/1000\n",
      "166868/166868 [==============================] - 2s 10us/sample - loss: 81.8921 - val_loss: 22.1091 - lr: 0.0030\n",
      "Epoch 90/1000\n",
      "166868/166868 [==============================] - 2s 10us/sample - loss: 81.6443 - val_loss: 22.1164 - lr: 0.0030\n",
      "Epoch 91/1000\n",
      "166868/166868 [==============================] - 2s 10us/sample - loss: 81.3962 - val_loss: 22.1419 - lr: 0.0030\n",
      "Epoch 92/1000\n",
      "166868/166868 [==============================] - 2s 10us/sample - loss: 81.1129 - val_loss: 22.1395 - lr: 0.0022\n",
      "Epoch 93/1000\n",
      "166868/166868 [==============================] - 2s 10us/sample - loss: 80.7719 - val_loss: 22.1515 - lr: 0.0022\n",
      "Epoch 94/1000\n",
      "166868/166868 [==============================] - 2s 9us/sample - loss: 80.6446 - val_loss: 22.1428 - lr: 0.0022\n",
      "Epoch 95/1000\n",
      "166868/166868 [==============================] - 2s 9us/sample - loss: 80.3898 - val_loss: 22.1667 - lr: 0.0022\n",
      "Epoch 96/1000\n",
      "166868/166868 [==============================] - 2s 10us/sample - loss: 80.2012 - val_loss: 22.1937 - lr: 0.0022\n",
      "Epoch 97/1000\n",
      "166868/166868 [==============================] - 2s 10us/sample - loss: 80.0563 - val_loss: 22.1867 - lr: 0.0022\n",
      "Epoch 98/1000\n",
      "166868/166868 [==============================] - 2s 9us/sample - loss: 79.8496 - val_loss: 22.1931 - lr: 0.0022\n",
      "Epoch 99/1000\n",
      "166868/166868 [==============================] - 2s 10us/sample - loss: 79.6680 - val_loss: 22.2054 - lr: 0.0022\n",
      "Epoch 100/1000\n",
      "166868/166868 [==============================] - 2s 10us/sample - loss: 79.4173 - val_loss: 22.2319 - lr: 0.0022\n",
      "Epoch 101/1000\n",
      "166868/166868 [==============================] - 2s 10us/sample - loss: 79.1576 - val_loss: 22.2375 - lr: 0.0022\n",
      "Epoch 102/1000\n",
      "166868/166868 [==============================] - 2s 10us/sample - loss: 78.9576 - val_loss: 22.2676 - lr: 0.0022\n",
      "Epoch 103/1000\n",
      "166868/166868 [==============================] - 2s 10us/sample - loss: 78.8389 - val_loss: 22.2739 - lr: 0.0022\n",
      "Epoch 104/1000\n",
      "166868/166868 [==============================] - 2s 10us/sample - loss: 78.5585 - val_loss: 22.2944 - lr: 0.0022\n",
      "Epoch 105/1000\n",
      "166868/166868 [==============================] - 2s 9us/sample - loss: 78.4250 - val_loss: 22.3146 - lr: 0.0022\n",
      "Epoch 106/1000\n",
      "166868/166868 [==============================] - 2s 10us/sample - loss: 78.2612 - val_loss: 22.3377 - lr: 0.0022\n",
      "Epoch 107/1000\n",
      "166868/166868 [==============================] - 2s 9us/sample - loss: 77.9752 - val_loss: 22.3413 - lr: 0.0022\n",
      "Epoch 108/1000\n",
      "166868/166868 [==============================] - 2s 10us/sample - loss: 77.8384 - val_loss: 22.3694 - lr: 0.0022\n",
      "Epoch 109/1000\n",
      "166868/166868 [==============================] - 2s 9us/sample - loss: 77.5605 - val_loss: 22.3750 - lr: 0.0022\n",
      "Epoch 110/1000\n",
      "166868/166868 [==============================] - 2s 10us/sample - loss: 77.3537 - val_loss: 22.3989 - lr: 0.0022\n",
      "Epoch 111/1000\n",
      "166868/166868 [==============================] - 2s 10us/sample - loss: 77.1860 - val_loss: 22.4170 - lr: 0.0017\n",
      "Epoch 112/1000\n",
      "166868/166868 [==============================] - 2s 10us/sample - loss: 77.0115 - val_loss: 22.4372 - lr: 0.0017\n",
      "Epoch 113/1000\n",
      "166868/166868 [==============================] - 2s 10us/sample - loss: 76.9459 - val_loss: 22.4525 - lr: 0.0017\n",
      "Epoch 114/1000\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "166868/166868 [==============================] - 2s 10us/sample - loss: 76.8082 - val_loss: 22.4556 - lr: 0.0017\n",
      "Epoch 115/1000\n",
      "166868/166868 [==============================] - 2s 10us/sample - loss: 76.6293 - val_loss: 22.4959 - lr: 0.0017\n",
      "Epoch 116/1000\n",
      "166868/166868 [==============================] - 2s 9us/sample - loss: 76.5126 - val_loss: 22.4986 - lr: 0.0017\n",
      "Epoch 117/1000\n",
      "166868/166868 [==============================] - 2s 10us/sample - loss: 76.2721 - val_loss: 22.5190 - lr: 0.0017\n",
      "Epoch 118/1000\n",
      "166868/166868 [==============================] - 2s 9us/sample - loss: 76.1296 - val_loss: 22.5483 - lr: 0.0017\n",
      "Epoch 119/1000\n",
      "166868/166868 [==============================] - 2s 9us/sample - loss: 76.0040 - val_loss: 22.5556 - lr: 0.0017\n",
      "Epoch 120/1000\n",
      "166868/166868 [==============================] - 2s 10us/sample - loss: 75.9725 - val_loss: 22.5776 - lr: 0.0017\n",
      "Epoch 121/1000\n",
      "166868/166868 [==============================] - 2s 9us/sample - loss: 75.7967 - val_loss: 22.5913 - lr: 0.0017\n",
      "Epoch 122/1000\n",
      "166868/166868 [==============================] - 2s 10us/sample - loss: 75.5896 - val_loss: 22.6054 - lr: 0.0017\n",
      "Epoch 123/1000\n",
      "166868/166868 [==============================] - 2s 10us/sample - loss: 75.4425 - val_loss: 22.6285 - lr: 0.0017\n",
      "Epoch 124/1000\n",
      "166868/166868 [==============================] - 2s 10us/sample - loss: 75.3231 - val_loss: 22.6227 - lr: 0.0017\n",
      "Epoch 125/1000\n",
      "166868/166868 [==============================] - 2s 10us/sample - loss: 75.1971 - val_loss: 22.6385 - lr: 0.0017\n",
      "Epoch 126/1000\n",
      "166868/166868 [==============================] - 2s 10us/sample - loss: 74.9964 - val_loss: 22.6414 - lr: 0.0017\n",
      "\n",
      "training finished in 126 epochs (early stop), transform data to adjust the platform effect...\n",
      "\n",
      "\n",
      "re-run DE on CVAE transformed scRNA-seq data!\n",
      "Differential analysis across cell-types on scRNA-seq data...\n",
      "finally selected 160 cell-type marker genes\n",
      "\n",
      "\n",
      "platform effect adjustment by CVAE finished. Elapsed time: 10.47 minutes.\n",
      "\n",
      "\n",
      "use the marker genes derived from CVAE transformed scRNA-seq for downstream regression!\n",
      "\n",
      "gene filtering before modeling...\n",
      "15 genes with nUMIs<5 in all spatial spots and need to be excluded\n",
      "finally use 145 genes for modeling\n",
      "\n",
      "spot filtering before modeling...\n",
      "all spots passed filtering\n",
      "\n",
      "\n",
      "######### Start GLRM modeling... #########\n",
      "\n",
      "GLRM settings:\n",
      "use SciPy minimize method:  L-BFGS-B\n",
      "global optimization turned off, local minimum will be used in GLRM\n",
      "use hybrid version of GLRM\n",
      "Numba detected total 64 available CPU cores. Use 32 CPU cores\n",
      "use 2001 points to calculate the heavy-tail density\n",
      "use weight threshold for Adaptive Lasso:  0.001\n",
      "total 164 unique nUMIs, min: 0.0, max: 285.0\n",
      "\n",
      "Build graph: \n",
      " Graph with 293 nodes and 534 edges\n",
      "\n",
      "estimation of gene-specific platform effect gamma_g is skipped as already using CVAE to adjust platform effect\n",
      "\n",
      "\n",
      "Start GLRM fitting...\n",
      "\n",
      "first estimate MLE theta and corresponding e^alpha and sigma^2...\n",
      "\n",
      "GLRM model initialization...\n",
      "calculate MLE theta and sigma^2 iteratively...\n",
      "  iter | time_opt | time_sig | sigma2\n",
      "     0 |    4.762 |    0.764 |  1.810\n",
      "     1 |    2.699 |    0.766 |  1.996\n",
      "     2 |    2.299 |    0.637 |  2.038\n",
      "     3 |    2.094 |    0.635 |  2.047\n",
      "     4 |    1.589 |    0.507 |  2.049\n",
      "     5 |    1.085 |    0.507 |  2.049\n",
      "MLE theta and sigma^2 calculation finished. Elapsed time: 0.31 minutes.\n",
      "MLE theta estimation finished. Elapsed time: 0.31 minutes.\n",
      "\n",
      "calculate weights of Adaptive Lasso...\n",
      "\n",
      "Stage 1: variable selection using Adaptive Lasso starts with the MLE theta and e^alpha, using already estimated sigma^2 and gamma_g...\n",
      "specified hyper-parameter for Adaptive Lasso is: [0.1, 0.268, 0.72, 1.931, 5.179, 13.895, 37.276, 100.0]\n",
      "hyper-parameter for Adaptive Lasso: use cross-validation to find the optimal value from 8 candidates...\n",
      "\n",
      "Start cross-validation for hyper-parameter lambda_r...\n",
      "directly estimate theta by Adaptive Lasso loss function as NO Graph Laplacian constrain!\n",
      "0%...11%...22%...33%...44%...56%...67%...78%...early stop\n",
      "find optimal lambda_r 1.931 with average negative log-likelihood 8473.4717 by 5 fold cross-validation. Elapsed time: 2.40 minutes.\n",
      "\n",
      "\n",
      "start ADMM iteration...\n",
      "  iter |  res_pri_n | res_dual_n |    eps_pri |   eps_dual |        rho |    new_rho | time_opt | time_reg | time_lap | tilde_RMSE |   hat_RMSE\n",
      "     0 |      9.623 |      9.623 |      0.074 |      0.074 |       1.00 |       1.00 |    0.812 |    0.000 |    0.003 |   0.293990 |   0.146995\n",
      "     1 |      9.573 |      0.162 |      0.074 |      0.083 |       1.00 |       2.00 |    2.008 |    0.000 |    0.002 |   0.292296 |   0.146148\n",
      "     2 |      9.437 |      0.550 |      0.073 |      0.102 |       2.00 |       4.00 |    2.004 |    0.000 |    0.002 |   0.288425 |   0.144212\n",
      "     3 |      6.892 |     13.463 |      0.078 |      0.129 |       4.00 |       4.00 |    2.277 |    0.000 |    0.002 |   0.213492 |   0.106746\n",
      "     4 |      4.860 |     15.719 |      0.080 |      0.144 |       4.00 |       4.00 |    2.017 |    0.000 |    0.002 |   0.123781 |   0.061890\n",
      "     5 |      4.461 |     10.809 |      0.075 |      0.153 |       4.00 |       4.00 |    2.150 |    0.000 |    0.003 |   0.127856 |   0.063928\n",
      "     6 |      4.160 |      6.657 |      0.071 |      0.160 |       4.00 |       4.00 |    2.133 |    0.000 |    0.002 |   0.122043 |   0.061022\n",
      "     7 |      3.572 |      5.858 |      0.070 |      0.168 |       4.00 |       8.00 |    2.179 |    0.000 |    0.002 |   0.099593 |   0.049797\n",
      "     8 |      3.020 |      8.972 |      0.073 |      0.186 |       8.00 |       8.00 |    2.154 |    0.000 |    0.002 |   0.077053 |   0.038527\n",
      "     9 |      2.612 |      9.289 |      0.073 |      0.203 |       8.00 |       8.00 |    2.102 |    0.000 |    0.002 |   0.066900 |   0.033450\n",
      "    10 |      2.282 |      8.218 |      0.072 |      0.216 |       8.00 |      16.00 |    1.980 |    0.000 |    0.002 |   0.060742 |   0.030371\n",
      "    11 |      1.964 |     10.796 |      0.075 |      0.241 |      16.00 |      16.00 |    2.080 |    0.000 |    0.002 |   0.051495 |   0.025747\n",
      "    12 |      1.704 |     11.768 |      0.076 |      0.261 |      16.00 |      16.00 |    2.072 |    0.000 |    0.002 |   0.043483 |   0.021741\n",
      "    13 |      1.501 |     10.430 |      0.074 |      0.277 |      16.00 |      32.00 |    2.091 |    0.000 |    0.002 |   0.038609 |   0.019305\n",
      "    14 |      1.259 |     13.709 |      0.078 |      0.307 |      32.00 |      32.00 |    2.202 |    0.000 |    0.002 |   0.032167 |   0.016084\n",
      "    15 |      1.044 |     15.705 |      0.080 |      0.331 |      32.00 |      32.00 |    2.101 |    0.000 |    0.002 |   0.026407 |   0.013204\n",
      "    16 |      0.912 |     14.179 |      0.078 |      0.350 |      32.00 |      64.00 |    2.226 |    0.000 |    0.002 |   0.023136 |   0.011568\n",
      "    17 |      0.766 |     16.839 |      0.081 |      0.384 |      64.00 |      64.00 |    2.260 |    0.000 |    0.002 |   0.019337 |   0.009669\n",
      "    18 |      0.646 |     18.866 |      0.083 |      0.413 |      64.00 |      64.00 |    2.183 |    0.000 |    0.002 |   0.015877 |   0.007939\n",
      "    19 |      0.550 |     17.873 |      0.082 |      0.436 |      64.00 |     128.00 |    2.290 |    0.000 |    0.002 |   0.013459 |   0.006730\n",
      "    20 |      0.462 |     21.546 |      0.086 |      0.477 |     128.00 |     128.00 |    2.197 |    0.000 |    0.002 |   0.011148 |   0.005574\n",
      "    21 |      0.390 |     23.064 |      0.087 |      0.511 |     128.00 |     128.00 |    2.302 |    0.000 |    0.002 |   0.009243 |   0.004621\n",
      "    22 |      0.339 |     20.707 |      0.085 |      0.539 |     128.00 |     256.00 |    2.180 |    0.000 |    0.002 |   0.007996 |   0.003998\n",
      "    23 |      0.275 |     26.453 |      0.090 |      0.586 |     256.00 |     256.00 |    2.118 |    0.000 |    0.002 |   0.006389 |   0.003195\n",
      "    24 |      0.228 |     28.388 |      0.092 |      0.624 |     256.00 |     256.00 |    2.245 |    0.000 |    0.002 |   0.005067 |   0.002534\n",
      "    25 |      0.199 |     24.676 |      0.089 |      0.654 |     256.00 |     512.00 |    2.191 |    0.000 |    0.002 |   0.004415 |   0.002208\n",
      "    26 |      0.167 |     29.232 |      0.093 |      0.708 |     512.00 |     512.00 |    2.183 |    0.000 |    0.002 |   0.003638 |   0.001819\n",
      "    27 |      0.143 |     31.549 |      0.096 |      0.756 |     512.00 |     512.00 |    2.167 |    0.000 |    0.002 |   0.002999 |   0.001499\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    28 |      0.126 |     28.083 |      0.092 |      0.797 |     512.00 |    1024.00 |    2.175 |    0.000 |    0.002 |   0.002587 |   0.001293\n",
      "    29 |      0.104 |     36.400 |      0.100 |      0.870 |    1024.00 |    1024.00 |    2.029 |    0.000 |    0.002 |   0.002117 |   0.001059\n",
      "    30 |      0.085 |     40.959 |      0.105 |      0.928 |    1024.00 |    1024.00 |    1.972 |    0.000 |    0.002 |   0.001659 |   0.000829\n",
      "    31 |      0.072 |     36.468 |      0.101 |      0.973 |    1024.00 |    2048.00 |    2.122 |    0.000 |    0.002 |   0.001373 |   0.000686\n",
      "    32 |      0.058 |     45.574 |      0.110 |      1.046 |    2048.00 |    2048.00 |    2.152 |    0.000 |    0.002 |   0.001079 |   0.000540\n",
      "    33 |      0.049 |     58.273 |      0.122 |      1.104 |    2048.00 |    2048.00 |    1.894 |    0.000 |    0.002 |   0.000853 |   0.000427\n",
      "    34 |      0.044 |     37.893 |      0.102 |      1.153 |    2048.00 |    4096.00 |    1.763 |    0.000 |    0.002 |   0.000729 |   0.000365\n",
      "    35 |      0.038 |     45.012 |      0.109 |      1.244 |    4096.00 |    4096.00 |    1.875 |    0.000 |    0.002 |   0.000591 |   0.000296\n",
      "    36 |      0.033 |     48.164 |      0.112 |      1.329 |    4096.00 |    4096.00 |    1.683 |    0.000 |    0.002 |   0.000473 |   0.000237\n",
      "    37 |      0.029 |     44.465 |      0.109 |      1.405 |    4096.00 |    8192.00 |    1.546 |    0.000 |    0.002 |   0.000395 |   0.000197\n",
      "    38 |      0.023 |    129.906 |      0.194 |      1.528 |    8192.00 |    8192.00 |    1.533 |    0.000 |    0.002 |   0.000307 |   0.000153\n",
      "    39 |      0.017 |     93.247 |      0.157 |      1.617 |    8192.00 |    8192.00 |    1.537 |    0.000 |    0.002 |   0.000226 |   0.000113\n",
      "    40 |      0.014 |     74.468 |      0.139 |      1.679 |    8192.00 |   16384.00 |    1.430 |    0.000 |    0.002 |   0.000185 |   0.000093\n",
      "    41 |      0.011 |     76.569 |      0.141 |      1.776 |   16384.00 |   16384.00 |    1.369 |    0.000 |    0.002 |   0.000139 |   0.000069\n",
      "    42 |      0.009 |     73.819 |      0.138 |      1.855 |   16384.00 |   16384.00 |    1.308 |    0.000 |    0.002 |   0.000102 |   0.000051\n",
      "    43 |      0.008 |     63.548 |      0.128 |      1.920 |   16384.00 |   32768.00 |    1.318 |    0.000 |    0.002 |   0.000082 |   0.000041\n",
      "    44 |      0.006 |     79.028 |      0.143 |      2.037 |   32768.00 |          / |    1.218 |    0.000 |    0.002 |   0.000062 |   0.000031\n",
      "early stop!\n",
      "Terminated (optimal) in 45 iterations.\n",
      "One optimization by ADMM finished. Elapsed time: 1.47 minutes.\n",
      "\n",
      "Stage 1 variable selection finished. Elapsed time: 3.87 minutes.\n",
      "\n",
      "Stage 2: final theta estimation with Graph Laplacian Constrain using already estimated sigma^2 and gamma_g\n",
      "specified hyper-parameter for Graph Laplacian Constrain is: [0.1, 0.268, 0.72, 1.931, 5.179, 13.895, 37.276, 100.0]\n",
      "hyper-parameter for Graph Laplacian Constrain: use cross-validation to find the optimal value from 8 candidates...\n",
      "\n",
      "Start cross-validation for hyper-parameter lambda_g...\n",
      "still use ADMM even NO Graph Laplacian constrain (lambda_g=0)\n",
      "0%...11%...22%...33%...44%...56%...67%...78%...89%...100%\n",
      "find optimal lambda_g 37.276 with average negative log-likelihood 8907.8486 by 5 fold cross-validation. Elapsed time: 20.54 minutes.\n",
      "\n",
      "\n",
      "start ADMM iteration...\n",
      "  iter |  res_pri_n | res_dual_n |    eps_pri |   eps_dual |        rho |    new_rho | time_opt | time_reg | time_lap | tilde_RMSE |   hat_RMSE\n",
      "     0 |     10.652 |     10.777 |      0.075 |      0.075 |       1.00 |       1.00 |    2.958 |    0.000 |    0.005 |   0.137222 |   0.180363\n",
      "     1 |      9.526 |      5.694 |      0.074 |      0.082 |       1.00 |       1.00 |    1.561 |    0.000 |    0.005 |   0.236476 |   0.182953\n",
      "     2 |      8.207 |      7.260 |      0.072 |      0.089 |       1.00 |       1.00 |    1.480 |    0.000 |    0.005 |   0.205195 |   0.160191\n",
      "     3 |      7.084 |      8.337 |      0.072 |      0.096 |       1.00 |       1.00 |    1.442 |    0.000 |    0.005 |   0.148043 |   0.141124\n",
      "     4 |      6.511 |      9.092 |      0.073 |      0.102 |       1.00 |       2.00 |    1.493 |    0.000 |    0.005 |   0.128574 |   0.128281\n",
      "     5 |      6.269 |     15.426 |      0.079 |      0.114 |       2.00 |       2.00 |    1.656 |    0.000 |    0.005 |   0.121130 |   0.122017\n",
      "     6 |      5.798 |     16.466 |      0.081 |      0.125 |       2.00 |       2.00 |    1.608 |    0.000 |    0.005 |   0.117564 |   0.110632\n",
      "     7 |      5.274 |     17.738 |      0.082 |      0.135 |       2.00 |       4.00 |    1.556 |    0.000 |    0.005 |   0.100082 |   0.100363\n",
      "     8 |      5.183 |     28.432 |      0.092 |      0.155 |       4.00 |       4.00 |    1.761 |    0.000 |    0.005 |   0.093136 |   0.096369\n",
      "     9 |      4.678 |     29.569 |      0.094 |      0.173 |       4.00 |       4.00 |    1.609 |    0.000 |    0.005 |   0.096025 |   0.085486\n",
      "    10 |      4.067 |     31.762 |      0.096 |      0.188 |       4.00 |       8.00 |    1.609 |    0.000 |    0.005 |   0.075696 |   0.075178\n",
      "    11 |      3.928 |     49.814 |      0.114 |      0.217 |       8.00 |       8.00 |    1.726 |    0.000 |    0.005 |   0.066956 |   0.070687\n",
      "    12 |      3.432 |     51.767 |      0.116 |      0.241 |       8.00 |       8.00 |    1.620 |    0.000 |    0.005 |   0.071257 |   0.060113\n",
      "    13 |      2.762 |     55.414 |      0.119 |      0.261 |       8.00 |      16.00 |    1.545 |    0.000 |    0.005 |   0.050295 |   0.049508\n",
      "    14 |      2.651 |     84.718 |      0.149 |      0.297 |      16.00 |      16.00 |    1.632 |    0.000 |    0.005 |   0.042453 |   0.046003\n",
      "    15 |      2.152 |     87.661 |      0.152 |      0.326 |      16.00 |      16.00 |    2.334 |    0.000 |    0.005 |   0.045876 |   0.036021\n",
      "    16 |      1.605 |     92.660 |      0.157 |      0.348 |      16.00 |      32.00 |    1.430 |    0.000 |    0.005 |   0.028020 |   0.027467\n",
      "    17 |      1.621 |    135.471 |      0.200 |      0.387 |      32.00 |      32.00 |    1.522 |    0.000 |    0.005 |   0.022709 |   0.025977\n",
      "    18 |      1.239 |    138.599 |      0.203 |      0.416 |      32.00 |      32.00 |    1.459 |    0.000 |    0.005 |   0.026475 |   0.018945\n",
      "    19 |      0.796 |    143.939 |      0.208 |      0.436 |      32.00 |      64.00 |    1.424 |    0.000 |    0.005 |   0.013286 |   0.012799\n",
      "    20 |      0.793 |    196.679 |      0.261 |      0.475 |      64.00 |      64.00 |    1.486 |    0.000 |    0.005 |   0.009663 |   0.012395\n",
      "    21 |      0.572 |    199.320 |      0.263 |      0.501 |      64.00 |      64.00 |    1.495 |    0.000 |    0.005 |   0.012918 |   0.008250\n",
      "    22 |      0.299 |    203.789 |      0.268 |      0.516 |      64.00 |     128.00 |    1.372 |    0.000 |    0.005 |   0.005088 |   0.004631\n",
      "    23 |      0.332 |    256.870 |      0.321 |      0.545 |     128.00 |     128.00 |    1.297 |    0.000 |    0.004 |   0.002972 |   0.004966\n",
      "    24 |      0.285 |    258.553 |      0.323 |      0.560 |     128.00 |     128.00 |    1.409 |    0.000 |    0.004 |   0.005555 |   0.003231\n",
      "    25 |      0.156 |    260.768 |      0.325 |      0.568 |     128.00 |     256.00 |    1.276 |    0.000 |    0.004 |   0.001907 |   0.001462\n",
      "    26 |      0.127 |    304.130 |      0.368 |      0.583 |     256.00 |     256.00 |    1.255 |    0.000 |    0.004 |   0.000720 |   0.001680\n",
      "    27 |      0.075 |    304.314 |      0.368 |      0.592 |     256.00 |     256.00 |    1.376 |    0.000 |    0.004 |   0.001765 |   0.000945\n",
      "    28 |      0.025 |    305.099 |      0.369 |      0.594 |     256.00 |     512.00 |    1.288 |    0.000 |    0.004 |   0.000489 |   0.000346\n",
      "    29 |      0.038 |    334.707 |      0.399 |      0.603 |     512.00 |     512.00 |    1.173 |    0.000 |    0.003 |   0.000142 |   0.000466\n",
      "    30 |      0.022 |    334.426 |      0.398 |      0.608 |     512.00 |     512.00 |    1.356 |    0.000 |    0.003 |   0.000520 |   0.000263\n",
      "    31 |      0.031 |    334.628 |      0.399 |      0.609 |     512.00 |    1024.00 |    1.258 |    0.000 |    0.003 |   0.000196 |   0.000113\n",
      "    32 |      0.011 |    352.399 |      0.416 |      0.614 |    1024.00 |    1024.00 |    1.150 |    0.000 |    0.003 |   0.000039 |   0.000131\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    33 |      0.006 |    352.072 |      0.416 |      0.618 |    1024.00 |    1024.00 |    1.259 |    0.000 |    0.003 |   0.000148 |   0.000074\n",
      "    34 |      0.002 |    351.852 |      0.416 |      0.618 |    1024.00 |          / |    1.135 |    0.000 |    0.003 |   0.000042 |   0.000023\n",
      "early stop!\n",
      "Terminated (optimal) in 35 iterations.\n",
      "One optimization by ADMM finished. Elapsed time: 0.89 minutes.\n",
      "\n",
      "\n",
      "stage 2 finished. Elapsed time: 21.43 minutes.\n",
      "\n",
      "GLRM fitting finished. Elapsed time: 25.61 minutes.\n",
      "\n",
      "\n",
      "Post-processing estimated cell-type proportion theta...\n",
      "hard thresholding small theta values with threshold 0\n",
      "\n",
      "\n",
      "cell type deconvolution finished. Estimate results saved in /home/exouser/Spatial/celltype_proportions.csv. Elapsed time: 0.60 hours.\n",
      "\n",
      "\n",
      "######### Start imputation #########\n",
      "imputation for 160 µm finished. Elapsed time: 0.35 minutes\n",
      "imputation for 114 µm finished. Elapsed time: 0.91 minutes\n",
      "imputation for 80 µm finished. Elapsed time: 2.88 minutes\n",
      "\n",
      "\n",
      "whole pipeline finished. Total elapsed time: 0.67 hours.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "CompletedProcess(args='runDeconvolution -q Melanoma_spatial_spot_nUMI.csv                           -r Melanoma_ref_scRNA_cell_nUMI.csv                           -c Melanoma_ref_scRNA_cell_celltype.csv                           -a Melanoma_spatial_spot_adjacency_matrix.csv                           -l Melanoma_spatial_spot_loc.csv                           --n_hv_gene 300                           --n_marker_per_cmp 10                           --pseudo_spot_min_cell 5                           --pseudo_spot_max_cell 40                           --seed 3                           -n 32                           --use_imputation true\\n', returncode=0)"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import subprocess\n",
    "\n",
    "cmd = '''runDeconvolution -q Melanoma_spatial_spot_nUMI.csv \\\n",
    "                          -r Melanoma_ref_scRNA_cell_nUMI.csv \\\n",
    "                          -c Melanoma_ref_scRNA_cell_celltype.csv \\\n",
    "                          -a Melanoma_spatial_spot_adjacency_matrix.csv \\\n",
    "                          -l Melanoma_spatial_spot_loc.csv \\\n",
    "                          --n_hv_gene 300 \\\n",
    "                          --n_marker_per_cmp 10 \\\n",
    "                          --pseudo_spot_min_cell 5 \\\n",
    "                          --pseudo_spot_max_cell 40 \\\n",
    "                          --seed 3 \\\n",
    "                          -n 32 \\\n",
    "                          --use_imputation true\n",
    "'''\n",
    "subprocess.run(cmd, check=True, text=True, shell=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a99aa164",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
