{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "db496b92",
   "metadata": {},
   "source": [
    "# Run *SDePER* on sequencing-based 3x simulated data: Scenario 1 + Spatial data as reference + NO CVAE\n",
    "\n",
    "In this Notebook we run SDePER on simulated data. For generating **sequencing-based high density** simulated data via coarse-graining procedure please refer [generate_simulated_spatial_data.nb.html](https://rawcdn.githack.com/az7jh2/SDePER_Analysis/d22e0c9b4574530a8ecbdf620638f2527ec20c5e/Simulation_seq_based/Generate_high_density_simulation_data/generate_simulated_spatial_data.nb.html) in [Generate_high_density_simulation_data](https://github.com/az7jh2/SDePER_Analysis/tree/main/Simulation_seq_based/Generate_high_density_simulation_data) folder.\n",
    "\n",
    "**Scenario 1** means the reference data for deconvolution includes all single cells with the **matched 12 cell types**.\n",
    "\n",
    "**Spatial data as reference** means the reference data is actually the [GSE102827](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE102827) scRNA-seq data which is used to generate the simulated data, therefore it's **free of platform effect**.\n",
    "\n",
    "**NO CVAE** means we DO NOT use CVAE to remove platform effect since it's free of platform effect here.\n",
    "\n",
    "==================================================================================================================\n",
    "\n",
    "So here we use the **4 input files** as shown below:\n",
    "\n",
    "1. raw nUMI counts of **3x** simulated spatial transcriptomic data (spots × genes): [sim_seq_based_3x_spatial_spot_nUMI.csv](https://github.com/az7jh2/SDePER_Analysis/blob/main/Simulation_seq_based/Generate_high_density_simulation_data/sim_seq_based_3x_spatial_spot_nUMI.csv)\n",
    "2. raw nUMI counts of reference GSE102827 scRNA-seq data (cells × genes): `GSE102827_scRNA_cell_nUMI.csv`. Since the file size of csv file of raw nUMI matrix of all 65,539 cells and 25,187 genes is up to 3.1 GB, we do not provide this file in our repository. It's just a **matrix transpose** of [GSE102827_merged_all_raw.csv.gz](https://www.ncbi.nlm.nih.gov/geo/download/?acc=GSE102827&format=file&file=GSE102827%5Fmerged%5Fall%5Fraw%2Ecsv%2Egz) in [GSE102827](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE102827) to satisty the file format requirement that rows as cells and columns as genes\n",
    "3. cell type annotations for selected **6,006 cells** used for simulated data generation in [GSE102827](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE102827) scRNA-seq data (cells × 1): [GSE102827_3x_scRNA_cell_celltype.csv](https://github.com/az7jh2/SDePER_Analysis/blob/main/Simulation_seq_based/Run_SDePER_on_high_density_simulation_data/Scenario_1/ref_spatial/GSE102827_3x_scRNA_cell_celltype.csv)\n",
    "4. adjacency matrix of spots in simulated spatial transcriptomic data (spots × spots): [sim_spatial_spot_adjacency_matrix.csv](https://github.com/az7jh2/SDePER_Analysis/blob/main/Simulation/Generate_simulation_data/sim_spatial_spot_adjacency_matrix.csv)\n",
    "\n",
    "==================================================================================================================\n",
    "\n",
    "SDePER settings are the same as those used for the 1x setting ([S1_ref_spatial_SDePER_NO_CVAE.ipynb](https://github.com/az7jh2/SDePER_Analysis/blob/main/Simulation_seq_based/Run_SDePER_on_simulation_data/Scenario_1/ref_spatial/S1_ref_spatial_SDePER_NO_CVAE.ipynb)):\n",
    "\n",
    "* number of selected TOP marker genes for each comparison in Differential `n_marker_per_cmp`: 20\n",
    "* number of used CPU cores `n_core`: 64\n",
    "* initial learning rate for training CVAE `cvae_init_lr`: 0.003\n",
    "* number of hidden layers in encoder and decoder of CVAE `num_hidden_layer`: 1\n",
    "* whether to use Batch Normalization `use_batch_norm`: false\n",
    "* CVAE training epochs `cvae_train_epoch`: 1000\n",
    "* for diagnostic purposes set `diagnosis`: true\n",
    "* **whether to use CVAE to remove platform effect `use_cvae`: false**\n",
    "\n",
    "ALL other options are left as default.\n",
    "\n",
    "==================================================================================================================\n",
    "\n",
    "the `bash` command to start cell type deconvolution is\n",
    "\n",
    "`runDeconvolution -q sim_seq_based_3x_spatial_spot_nUMI.csv -r GSE102827_scRNA_cell_nUMI.csv -c GSE102827_3x_scRNA_cell_celltype.csv -a sim_spatial_spot_adjacency_matrix.csv --n_marker_per_cmp 20 -n 64 --cvae_init_lr 0.003 --num_hidden_layer 1 --use_batch_norm false --cvae_train_epoch 1000 --diagnosis true --use_cvae false`\n",
    "\n",
    "Note this Notebook uses **SDePER v1.2.1**. Cell type deconvolution result is renamed as [3x_S1_ref_spatial_SDePER_NO_CVAE_celltype_proportions.csv](https://github.com/az7jh2/SDePER_Analysis/blob/main/Simulation_seq_based/Run_SDePER_on_high_density_simulation_data/Scenario_1/ref_spatial/3x_S1_ref_spatial_SDePER_NO_CVAE_celltype_proportions.csv)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "df9e3939",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "SDePER (Spatial Deconvolution method with Platform Effect Removal) v1.2.1\n",
      "\n",
      "\n",
      "running options:\n",
      "spatial_file: /home/exouser/Spatial/sim_seq_based_3x_spatial_spot_nUMI.csv\n",
      "ref_file: /home/exouser/Spatial/GSE102827_scRNA_cell_nUMI.csv\n",
      "ref_celltype_file: /home/exouser/Spatial/GSE102827_3x_scRNA_cell_celltype.csv\n",
      "marker_file: None\n",
      "loc_file: None\n",
      "A_file: /home/exouser/Spatial/sim_spatial_spot_adjacency_matrix.csv\n",
      "n_cores: 64\n",
      "threshold: 0\n",
      "use_cvae: False\n",
      "use_imputation: False\n",
      "diagnosis: True\n",
      "verbose: True\n",
      "use_fdr: True\n",
      "p_val_cutoff: 0.05\n",
      "fc_cutoff: 1.2\n",
      "pct1_cutoff: 0.3\n",
      "pct2_cutoff: 0.1\n",
      "sortby_fc: True\n",
      "n_marker_per_cmp: 20\n",
      "filter_cell: True\n",
      "filter_gene: True\n",
      "n_hv_gene: 200\n",
      "n_pseudo_spot: 500000\n",
      "pseudo_spot_min_cell: 2\n",
      "pseudo_spot_max_cell: 8\n",
      "seq_depth_scaler: 10000\n",
      "cvae_input_scaler: 10\n",
      "cvae_init_lr: 0.003\n",
      "num_hidden_layer: 1\n",
      "use_batch_norm: False\n",
      "cvae_train_epoch: 1000\n",
      "use_spatial_pseudo: False\n",
      "redo_de: True\n",
      "seed: 383\n",
      "lambda_r: [0.1, 0.268, 0.72, 1.931, 5.179, 13.895, 37.276, 100.0]\n",
      "lambda_g: [0.1, 0.268, 0.72, 1.931, 5.179, 13.895, 37.276, 100.0]\n",
      "diameter: 200\n",
      "impute_diameter: [160, 114, 80]\n",
      "\n",
      "\n",
      "######### Preprocessing... #########\n",
      "\n",
      "building CVAE skipped...\n",
      "\n",
      "read spatial data from file /home/exouser/Spatial/sim_seq_based_3x_spatial_spot_nUMI.csv\n",
      "total 581 spots; 25187 genes\n",
      "\n",
      "filtering genes present in <3 spots: 8007 genes removed\n",
      "\n",
      "no marker gene profile provided. Perform DE to get cell-type marker genes on scRNA-seq data...\n",
      "\n",
      "read scRNA-seq data from file /home/exouser/Spatial/GSE102827_scRNA_cell_nUMI.csv\n",
      "total 65539 cells; 25187 genes\n",
      "read scRNA-seq cell-type annotation from file /home/exouser/Spatial/GSE102827_3x_scRNA_cell_celltype.csv\n",
      "total 12 cell-types\n",
      "subset cells with cell-type annotation, finally keep 6006 cells; 25187 genes\n",
      "\n",
      "filtering cells with <200 genes: No cells removed\n",
      "\n",
      "filtering genes present in <10 cells: 9756 genes removed\n",
      "\n",
      "Differential analysis across cell-types on scRNA-seq data...\n",
      "0%...8%...WARNING: only 0 genes passing filtering (<20) for Endo vs Smc\n",
      "17%...25%...33%...42%...50%...58%...67%...WARNING: only 19 genes passing filtering (<20) for eL2/3 vs eL4\n",
      "75%...WARNING: only 14 genes passing filtering (<20) for eL4 vs eL2/3\n",
      "WARNING: only 11 genes passing filtering (<20) for eL4 vs eL5\n",
      "83%...WARNING: only 18 genes passing filtering (<20) for eL5 vs eL2/3\n",
      "WARNING: only 15 genes passing filtering (<20) for eL5 vs eL4\n",
      "92%...WARNING: only 11 genes passing filtering (<20) for eL6 vs eL5\n",
      "finally selected 518 cell-type marker genes\n",
      "\n",
      "\n",
      "gene filtering before modeling...\n",
      "all genes passed filtering\n",
      "\n",
      "spot filtering before modeling...\n",
      "all spots passed filtering\n",
      "\n",
      "\n",
      "######### Start GLRM modeling... #########\n",
      "\n",
      "GLRM settings:\n",
      "use SciPy minimize method:  L-BFGS-B\n",
      "global optimization turned off, local minimum will be used in GLRM\n",
      "use hybrid version of GLRM\n",
      "Numba detected total 64 available CPU cores. Use 64 CPU cores\n",
      "use 2001 points to calculate the heavy-tail density\n",
      "use weight threshold for Adaptive Lasso:  0.001\n",
      "total 26 unique nUMIs, min: 0, max: 28\n",
      "\n",
      "Build graph: \n",
      " Graph with 581 nodes and 1029 edges\n",
      "\n",
      "Estimate gene-specific platform effect gamma_g...\n",
      "estimate platform effect gammg_g by L-BFGS-B and basinhopping\n",
      "\n",
      "GLRM model initialization...\n",
      "CAUTION: global optimization turned on!\n",
      "calculate MLE theta and sigma^2 iteratively...\n",
      "  iter | time_opt | time_sig | sigma2\n",
      "     0 |    1.748 |    0.151 |  0.574\n",
      "     1 |    0.465 |    0.146 |  0.574\n",
      "MLE theta and sigma^2 calculation finished. Elapsed time: 0.04 minutes.\n",
      "gamma_g estimation finished. Elapsed time: 0.04 minutes.\n",
      "\n",
      "\n",
      "Start GLRM fitting...\n",
      "\n",
      "first estimate MLE theta and corresponding e^alpha and sigma^2...\n",
      "\n",
      "GLRM model initialization...\n",
      "calculate MLE theta and sigma^2 iteratively...\n",
      "  iter | time_opt | time_sig | sigma2\n",
      "     0 |   33.824 |    0.836 |  0.010\n",
      "     1 |   23.937 |    0.418 |  0.010\n",
      "MLE theta and sigma^2 calculation finished. Elapsed time: 0.98 minutes.\n",
      "MLE theta estimation finished. Elapsed time: 0.98 minutes.\n",
      "\n",
      "calculate weights of Adaptive Lasso...\n",
      "\n",
      "Stage 1: variable selection using Adaptive Lasso starts with the MLE theta and e^alpha, using already estimated sigma^2 and gamma_g...\n",
      "specified hyper-parameter for Adaptive Lasso is: [0.1, 0.268, 0.72, 1.931, 5.179, 13.895, 37.276, 100.0]\n",
      "hyper-parameter for Adaptive Lasso: use cross-validation to find the optimal value from 8 candidates...\n",
      "\n",
      "Start cross-validation for hyper-parameter lambda_r...\n",
      "directly estimate theta by Adaptive Lasso loss function as NO Graph Laplacian constrain!\n",
      "0%...11%...22%...33%...44%...56%...early stop\n",
      "find optimal lambda_r 0.268 with average negative log-likelihood 49072.4977 by 5 fold cross-validation. Elapsed time: 9.39 minutes.\n",
      "\n",
      "\n",
      "start ADMM iteration...\n",
      "  iter |  res_pri_n | res_dual_n |    eps_pri |   eps_dual |        rho |    new_rho | time_opt | time_reg | time_lap | tilde_RMSE |   hat_RMSE\n",
      "     0 |      7.950 |      7.950 |      0.126 |      0.126 |       1.00 |       1.00 |    2.733 |    0.000 |    0.003 |   0.133087 |   0.066544\n",
      "     1 |      7.933 |      0.035 |      0.126 |      0.134 |       1.00 |       2.00 |   10.375 |    0.000 |    0.003 |   0.132816 |   0.066408\n",
      "     2 |      6.495 |      4.767 |      0.125 |      0.147 |       2.00 |       2.00 |   11.363 |    0.000 |    0.003 |   0.106625 |   0.053312\n",
      "     3 |      5.101 |      6.094 |      0.124 |      0.156 |       2.00 |       4.00 |    9.939 |    0.000 |    0.003 |   0.080725 |   0.040362\n",
      "     4 |      4.325 |      7.360 |      0.125 |      0.170 |       4.00 |       4.00 |   13.065 |    0.000 |    0.003 |   0.069535 |   0.034768\n",
      "     5 |      3.728 |      7.092 |      0.125 |      0.181 |       4.00 |       4.00 |   11.146 |    0.000 |    0.003 |   0.060407 |   0.030204\n",
      "     6 |      3.383 |      5.579 |      0.124 |      0.190 |       4.00 |       8.00 |   10.969 |    0.000 |    0.003 |   0.054823 |   0.027411\n",
      "     7 |      2.956 |      7.363 |      0.125 |      0.208 |       8.00 |       8.00 |   11.806 |    0.000 |    0.003 |   0.047956 |   0.023978\n",
      "     8 |      2.560 |      8.451 |      0.127 |      0.224 |       8.00 |       8.00 |   10.841 |    0.000 |    0.003 |   0.041464 |   0.020732\n",
      "     9 |      2.320 |      7.275 |      0.125 |      0.238 |       8.00 |      16.00 |   10.486 |    0.000 |    0.003 |   0.037544 |   0.018772\n",
      "    10 |      1.999 |     10.317 |      0.128 |      0.263 |      16.00 |      16.00 |   12.118 |    0.000 |    0.003 |   0.032391 |   0.016196\n",
      "    11 |      1.678 |     12.296 |      0.130 |      0.283 |      16.00 |      16.00 |   11.316 |    0.000 |    0.003 |   0.027045 |   0.013523\n",
      "    12 |      1.470 |     10.720 |      0.129 |      0.299 |      16.00 |      32.00 |   14.187 |    0.000 |    0.003 |   0.023749 |   0.011875\n",
      "    13 |      1.238 |     13.929 |      0.132 |      0.326 |      32.00 |      32.00 |   12.523 |    0.000 |    0.003 |   0.020027 |   0.010013\n",
      "    14 |      1.034 |     15.452 |      0.134 |      0.348 |      32.00 |      32.00 |   11.649 |    0.000 |    0.003 |   0.016669 |   0.008334\n",
      "    15 |      0.922 |     12.822 |      0.131 |      0.366 |      32.00 |      64.00 |   10.735 |    0.000 |    0.003 |   0.014910 |   0.007455\n",
      "    16 |      0.807 |     15.097 |      0.133 |      0.400 |      64.00 |      64.00 |   11.254 |    0.000 |    0.003 |   0.013090 |   0.006545\n",
      "    17 |      0.712 |     16.619 |      0.135 |      0.432 |      64.00 |      64.00 |   10.321 |    0.000 |    0.003 |   0.011568 |   0.005784\n",
      "    18 |      0.650 |     14.467 |      0.133 |      0.461 |      64.00 |     128.00 |    9.753 |    0.000 |    0.003 |   0.010579 |   0.005290\n",
      "    19 |      0.587 |     18.377 |      0.136 |      0.517 |     128.00 |     128.00 |   10.590 |    0.000 |    0.003 |   0.009535 |   0.004767\n",
      "    20 |      0.525 |     21.531 |      0.140 |      0.569 |     128.00 |     128.00 |    9.206 |    0.000 |    0.003 |   0.008516 |   0.004258\n",
      "    21 |      0.474 |     21.207 |      0.139 |      0.613 |     128.00 |     256.00 |    8.840 |    0.000 |    0.003 |   0.007665 |   0.003832\n",
      "    22 |      0.416 |     31.480 |      0.150 |      0.694 |     256.00 |     256.00 |    8.758 |    0.000 |    0.003 |   0.006700 |   0.003350\n",
      "    23 |      0.363 |     35.863 |      0.154 |      0.762 |     256.00 |     256.00 |    8.362 |    0.000 |    0.003 |   0.005786 |   0.002893\n",
      "    24 |      0.326 |     32.208 |      0.150 |      0.819 |     256.00 |     512.00 |    8.968 |    0.000 |    0.003 |   0.005180 |   0.002590\n",
      "    25 |      0.281 |     44.677 |      0.163 |      0.921 |     512.00 |     512.00 |    7.987 |    0.000 |    0.003 |   0.004440 |   0.002220\n",
      "    26 |      0.237 |     52.468 |      0.171 |      1.006 |     512.00 |     512.00 |    7.785 |    0.000 |    0.003 |   0.003690 |   0.001845\n",
      "    27 |      0.207 |     47.208 |      0.165 |      1.072 |     512.00 |    1024.00 |    7.339 |    0.000 |    0.003 |   0.003215 |   0.001608\n",
      "    28 |      0.174 |     61.305 |      0.179 |      1.189 |    1024.00 |    1024.00 |    7.371 |    0.000 |    0.003 |   0.002682 |   0.001341\n",
      "    29 |      0.142 |     70.331 |      0.188 |      1.283 |    1024.00 |    1024.00 |    6.863 |    0.000 |    0.003 |   0.002141 |   0.001070\n",
      "    30 |      0.121 |     61.783 |      0.180 |      1.352 |    1024.00 |    2048.00 |    6.600 |    0.000 |    0.003 |   0.001790 |   0.000895\n",
      "    31 |      0.096 |     80.510 |      0.199 |      1.468 |    2048.00 |    2048.00 |    6.843 |    0.000 |    0.003 |   0.001395 |   0.000698\n",
      "    32 |      0.075 |     86.629 |      0.205 |      1.553 |    2048.00 |    2048.00 |    6.498 |    0.000 |    0.003 |   0.001043 |   0.000521\n",
      "    33 |      0.063 |     71.475 |      0.190 |      1.612 |    2048.00 |    4096.00 |    6.154 |    0.000 |    0.003 |   0.000854 |   0.000427\n",
      "    34 |      0.050 |     83.619 |      0.202 |      1.715 |    4096.00 |    4096.00 |    6.096 |    0.000 |    0.003 |   0.000648 |   0.000324\n",
      "    35 |      0.038 |     91.434 |      0.210 |      1.793 |    4096.00 |    4096.00 |    5.902 |    0.000 |    0.003 |   0.000463 |   0.000231\n",
      "    36 |      0.031 |     76.641 |      0.195 |      1.847 |    4096.00 |    8192.00 |    5.275 |    0.000 |    0.003 |   0.000370 |   0.000185\n",
      "    37 |      0.025 |     83.328 |      0.201 |      1.939 |    8192.00 |    8192.00 |    5.095 |    0.000 |    0.003 |   0.000276 |   0.000138\n",
      "    38 |      0.019 |     88.803 |      0.207 |      2.007 |    8192.00 |    8192.00 |    4.853 |    0.000 |    0.003 |   0.000191 |   0.000095\n",
      "    39 |      0.016 |     74.866 |      0.193 |      2.057 |    8192.00 |   16384.00 |    4.411 |    0.000 |    0.003 |   0.000152 |   0.000076\n",
      "    40 |      0.013 |     76.782 |      0.195 |      2.146 |   16384.00 |   16384.00 |    4.231 |    0.000 |    0.003 |   0.000116 |   0.000058\n",
      "    41 |      0.010 |     84.263 |      0.202 |      2.218 |   16384.00 |   16384.00 |    3.958 |    0.000 |    0.003 |   0.000082 |   0.000041\n",
      "    42 |      0.008 |     74.827 |      0.193 |      2.270 |   16384.00 |          / |    3.874 |    0.000 |    0.003 |   0.000062 |   0.000031\n",
      "early stop!\n",
      "Terminated (optimal) in 43 iterations.\n",
      "One optimization by ADMM finished. Elapsed time: 6.16 minutes.\n",
      "\n",
      "Stage 1 variable selection finished. Elapsed time: 15.55 minutes.\n",
      "\n",
      "Stage 2: final theta estimation with Graph Laplacian Constrain using already estimated sigma^2 and gamma_g\n",
      "specified hyper-parameter for Graph Laplacian Constrain is: [0.1, 0.268, 0.72, 1.931, 5.179, 13.895, 37.276, 100.0]\n",
      "hyper-parameter for Graph Laplacian Constrain: use cross-validation to find the optimal value from 8 candidates...\n",
      "\n",
      "Start cross-validation for hyper-parameter lambda_g...\n",
      "still use ADMM even NO Graph Laplacian constrain (lambda_g=0)\n",
      "0%...11%...22%...33%...44%...56%...67%...78%...89%...early stop\n",
      "find optimal lambda_g 5.179 with average negative log-likelihood 49054.3601 by 5 fold cross-validation. Elapsed time: 118.29 minutes.\n",
      "\n",
      "\n",
      "start ADMM iteration...\n",
      "  iter |  res_pri_n | res_dual_n |    eps_pri |   eps_dual |        rho |    new_rho | time_opt | time_reg | time_lap | tilde_RMSE |   hat_RMSE\n",
      "     0 |      9.803 |      7.218 |      0.128 |      0.128 |       1.00 |       1.00 |   29.425 |    0.000 |    0.007 |   0.124452 |   0.095691\n",
      "     1 |      7.463 |      5.954 |      0.126 |      0.134 |       1.00 |       1.00 |    7.086 |    0.000 |    0.007 |   0.084397 |   0.074454\n",
      "     2 |      6.596 |      5.341 |      0.125 |      0.140 |       1.00 |       1.00 |    8.035 |    0.000 |    0.007 |   0.071063 |   0.066708\n",
      "     3 |      6.111 |      5.136 |      0.124 |      0.145 |       1.00 |       2.00 |    7.468 |    0.000 |    0.007 |   0.066817 |   0.062249\n",
      "     4 |      5.378 |      8.515 |      0.127 |      0.155 |       2.00 |       2.00 |    9.444 |    0.000 |    0.007 |   0.060415 |   0.054887\n",
      "     5 |      4.662 |      8.537 |      0.127 |      0.165 |       2.00 |       2.00 |    8.666 |    0.000 |    0.008 |   0.049236 |   0.048352\n",
      "     6 |      4.129 |      8.725 |      0.127 |      0.173 |       2.00 |       4.00 |    8.582 |    0.000 |    0.007 |   0.043122 |   0.043149\n",
      "     7 |      3.541 |     13.747 |      0.132 |      0.186 |       4.00 |       4.00 |   10.050 |    0.000 |    0.007 |   0.040355 |   0.036749\n",
      "     8 |      2.912 |     14.110 |      0.132 |      0.197 |       4.00 |       4.00 |    9.143 |    0.000 |    0.007 |   0.031180 |   0.030274\n",
      "     9 |      2.382 |     14.587 |      0.133 |      0.207 |       4.00 |       8.00 |    8.579 |    0.000 |    0.007 |   0.024206 |   0.024913\n",
      "    10 |      1.895 |     21.272 |      0.139 |      0.221 |       8.00 |       8.00 |    9.962 |    0.000 |    0.007 |   0.022001 |   0.019486\n",
      "    11 |      1.386 |     21.563 |      0.140 |      0.231 |       8.00 |       8.00 |    8.808 |    0.000 |    0.007 |   0.014823 |   0.014343\n",
      "    12 |      0.999 |     21.883 |      0.140 |      0.238 |       8.00 |      16.00 |    7.875 |    0.000 |    0.007 |   0.009552 |   0.010372\n",
      "    13 |      0.716 |     28.833 |      0.147 |      0.248 |      16.00 |      16.00 |    9.279 |    0.000 |    0.006 |   0.008240 |   0.007279\n",
      "    14 |      0.444 |     28.958 |      0.147 |      0.254 |      16.00 |      16.00 |    7.102 |    0.000 |    0.006 |   0.004851 |   0.004494\n",
      "    15 |      0.256 |     29.099 |      0.147 |      0.257 |      16.00 |      32.00 |    4.665 |    0.000 |    0.006 |   0.002071 |   0.002527\n",
      "    16 |      0.168 |     34.674 |      0.153 |      0.261 |      32.00 |      32.00 |    6.506 |    0.000 |    0.005 |   0.001548 |   0.001641\n",
      "    17 |      0.097 |     34.731 |      0.153 |      0.262 |      32.00 |      32.00 |    4.597 |    0.000 |    0.005 |   0.001193 |   0.000843\n",
      "    18 |      0.046 |     34.850 |      0.153 |      0.263 |      32.00 |      64.00 |    3.777 |    0.000 |    0.005 |   0.000508 |   0.000314\n",
      "    19 |      0.042 |     38.700 |      0.157 |      0.263 |      64.00 |      64.00 |    3.312 |    0.000 |    0.004 |   0.000229 |   0.000314\n",
      "    20 |      0.026 |     38.745 |      0.157 |      0.263 |      64.00 |      64.00 |    3.678 |    0.000 |    0.004 |   0.000375 |   0.000196\n",
      "    21 |      0.010 |     38.816 |      0.157 |      0.263 |      64.00 |     128.00 |    3.505 |    0.000 |    0.004 |   0.000147 |   0.000082\n",
      "    22 |      0.012 |     41.183 |      0.159 |      0.263 |     128.00 |     128.00 |    3.145 |    0.000 |    0.004 |   0.000065 |   0.000084\n",
      "    23 |      0.008 |     41.189 |      0.159 |      0.263 |     128.00 |     128.00 |    3.193 |    0.000 |    0.004 |   0.000109 |   0.000056\n",
      "    24 |      0.004 |     41.205 |      0.159 |      0.263 |     128.00 |          / |    3.290 |    0.000 |    0.004 |   0.000056 |   0.000029\n",
      "early stop!\n",
      "Terminated (optimal) in 25 iterations.\n",
      "One optimization by ADMM finished. Elapsed time: 3.17 minutes.\n",
      "\n",
      "\n",
      "stage 2 finished. Elapsed time: 121.46 minutes.\n",
      "\n",
      "GLRM fitting finished. Elapsed time: 137.99 minutes.\n",
      "\n",
      "\n",
      "Post-processing estimated cell-type proportion theta...\n",
      "hard thresholding small theta values with threshold 0\n",
      "\n",
      "\n",
      "cell type deconvolution finished. Estimate results saved in /home/exouser/Spatial/celltype_proportions.csv. Elapsed time: 2.42 hours.\n",
      "\n",
      "\n",
      "######### No imputation #########\n",
      "\n",
      "\n",
      "whole pipeline finished. Total elapsed time: 2.42 hours.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "CompletedProcess(args='runDeconvolution -q sim_seq_based_3x_spatial_spot_nUMI.csv                           -r GSE102827_scRNA_cell_nUMI.csv                           -c GSE102827_3x_scRNA_cell_celltype.csv                           -a sim_spatial_spot_adjacency_matrix.csv                           --n_marker_per_cmp 20                           -n 64                           --cvae_init_lr 0.003                           --num_hidden_layer 1                           --use_batch_norm false                           --cvae_train_epoch 1000                           --diagnosis true                           --use_cvae false\\n', returncode=0)"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import subprocess\n",
    "\n",
    "cmd = '''runDeconvolution -q sim_seq_based_3x_spatial_spot_nUMI.csv \\\n",
    "                          -r GSE102827_scRNA_cell_nUMI.csv \\\n",
    "                          -c GSE102827_3x_scRNA_cell_celltype.csv \\\n",
    "                          -a sim_spatial_spot_adjacency_matrix.csv \\\n",
    "                          --n_marker_per_cmp 20 \\\n",
    "                          -n 64 \\\n",
    "                          --cvae_init_lr 0.003 \\\n",
    "                          --num_hidden_layer 1 \\\n",
    "                          --use_batch_norm false \\\n",
    "                          --cvae_train_epoch 1000 \\\n",
    "                          --diagnosis true \\\n",
    "                          --use_cvae false\n",
    "'''\n",
    "\n",
    "subprocess.run(cmd, check=True, text=True, shell=True)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
